Index: source/App/TAppDecoder/TAppDecTop.cpp
===================================================================
--- source/App/TAppDecoder/TAppDecTop.cpp	(revision 2764)
+++ source/App/TAppDecoder/TAppDecTop.cpp	(working copy)
@@ -172,7 +172,9 @@
     }
     if (bNewPicture || !bitstreamFile)
     {
-      m_cTDecTop.executeDeblockAndAlf(uiPOC, pcListPic, m_iSkipFrame, m_iPOCLastDisplay);
+      TComList<TComPic*> *rpcListPic = m_cTDecTop.executeDeblockAndAlf(uiPOC, m_iSkipFrame, m_iPOCLastDisplay);
+      if (rpcListPic)
+        pcListPic = rpcListPic;
     }
 
     if( pcListPic )
Index: source/Lib/TLibDecoder/TDecSlice.cpp
===================================================================
--- source/Lib/TLibDecoder/TDecSlice.cpp	(revision 2764)
+++ source/Lib/TLibDecoder/TDecSlice.cpp	(working copy)
@@ -90,7 +90,7 @@
   m_pcCuDecoder       = pcCuDecoder;
 }
 
-Void TDecSlice::decompressSlice(TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams, TComPic*& rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders)
+Void TDecSlice::decompressSlice(TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams, TComPic* rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders)
 {
   TComDataCU* pcCU;
   UInt        uiIsLast = 0;
Index: source/Lib/TLibDecoder/TDecBinCoderCABAC.cpp
===================================================================
--- source/Lib/TLibDecoder/TDecBinCoderCABAC.cpp	(revision 2764)
+++ source/Lib/TLibDecoder/TDecBinCoderCABAC.cpp	(working copy)
@@ -105,7 +105,7 @@
 Void
 TDecBinCABAC::decodeBin( UInt& ruiBin, ContextModel &rcCtxModel )
 {
-  UInt uiLPS = TComCABACTables::sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) - 4 ];
+  UInt uiLPS = sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) - 4 ];
   m_uiRange -= uiLPS;
   UInt scaledRange = m_uiRange << 7;
   
@@ -132,7 +132,7 @@
   else
   {
     // LPS path
-    Int numBits = TComCABACTables::sm_aucRenormTable[ uiLPS >> 3 ];
+    Int numBits = sm_aucRenormTable[ uiLPS >> 3 ];
     m_uiValue   = ( m_uiValue - scaledRange ) << numBits;
     m_uiRange   = uiLPS << numBits;
     ruiBin      = 1 - rcCtxModel.getMps();
Index: source/Lib/TLibDecoder/TDecEntropy.cpp
===================================================================
--- source/Lib/TLibDecoder/TDecEntropy.cpp	(revision 2764)
+++ source/Lib/TLibDecoder/TDecEntropy.cpp	(working copy)
@@ -290,7 +290,7 @@
 
 Void TDecEntropy::decodeMVPIdxPU( TComDataCU* pcSubCU, UInt uiPartAddr, UInt uiDepth, UInt uiPartIdx, RefPicList eRefList )
 {
-  Int iMVPIdx = -1;
+  Int iMVPIdx = 255;
 
   TComMv cZeroMv( 0, 0 );
   TComMv cMv     = cZeroMv;
@@ -311,7 +311,7 @@
   pcSubCU->setMVPIdxSubParts( iMVPIdx, eRefList, uiPartAddr, uiPartIdx, uiDepth );
   if ( iRefIdx >= 0 )
   {
-    m_pcPrediction->getMvPredAMVP( pcSubCU, uiPartIdx, uiPartAddr, eRefList, iRefIdx, cMv);
+    cMv = m_pcPrediction->getMvPredAMVP( pcSubCU, uiPartIdx, uiPartAddr, eRefList, iRefIdx);
     cMv += pcSubCUMvField->getMvd( uiPartAddr );
   }
 
Index: source/Lib/TLibDecoder/TDecTop.cpp
===================================================================
--- source/Lib/TLibDecoder/TDecTop.cpp	(revision 2764)
+++ source/Lib/TLibDecoder/TDecTop.cpp	(working copy)
@@ -191,14 +191,15 @@
 #endif
 }
 
-Void TDecTop::executeDeblockAndAlf(UInt& ruiPOC, TComList<TComPic*>*& rpcListPic, Int& iSkipFrame, Int& iPOCLastDisplay)
+TComList<TComPic*> *TDecTop::executeDeblockAndAlf(UInt& ruiPOC, Int& iSkipFrame, Int& iPOCLastDisplay)
 {
   if (!m_pcPic)
   {
     /* nothing to deblock */
-    return;
+    return NULL;
   }
-  
+
+  TComList<TComPic*>* rpcListPic = NULL;
   TComPic*&   pcPic         = m_pcPic;
 
   // Execute Deblock and ALF only + Cleanup
@@ -211,7 +212,7 @@
   m_cCuDecoder.destroy();        
   m_bFirstSliceInPicture  = true;
 
-  return;
+  return rpcListPic;
 }
 
 Void TDecTop::xCreateLostPicture(Int iLostPoc) 
Index: source/Lib/TLibDecoder/TDecSlice.h
===================================================================
--- source/Lib/TLibDecoder/TDecSlice.h	(revision 2764)
+++ source/Lib/TLibDecoder/TDecSlice.h	(working copy)
@@ -79,7 +79,7 @@
   Void  create            ( TComSlice* pcSlice, Int iWidth, Int iHeight, UInt uiMaxWidth, UInt uiMaxHeight, UInt uiMaxDepth );
   Void  destroy           ();
   
-  Void  decompressSlice   ( TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams,   TComPic*& rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders );
+  Void  decompressSlice   ( TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams,   TComPic* rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders );
 };
 
 
Index: source/Lib/TLibDecoder/TDecEntropy.h
===================================================================
--- source/Lib/TLibDecoder/TDecEntropy.h	(revision 2764)
+++ source/Lib/TLibDecoder/TDecEntropy.h	(working copy)
@@ -75,7 +75,7 @@
 #endif
   virtual void parseSEI(SEImessages&) = 0;
 
-  virtual Void parseSliceHeader          ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager)       = 0;
+  virtual Void parseSliceHeader          ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager)       = 0;
 
   virtual Void  parseTerminatingBit       ( UInt& ruilsLast )                                     = 0;
   
@@ -149,7 +149,7 @@
 #endif
   void decodeSEI(SEImessages& seis) { m_pcEntropyDecoderIf->parseSEI(seis); }
 
-  Void    decodeSliceHeader           ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager)  { m_pcEntropyDecoderIf->parseSliceHeader(rpcSlice, parameterSetManager);         }
+  Void    decodeSliceHeader           ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager)  { m_pcEntropyDecoderIf->parseSliceHeader(rpcSlice, parameterSetManager);         }
 
   Void    decodeTerminatingBit        ( UInt& ruiIsLast )       { m_pcEntropyDecoderIf->parseTerminatingBit(ruiIsLast);     }
   
Index: source/Lib/TLibDecoder/TDecTop.h
===================================================================
--- source/Lib/TLibDecoder/TDecTop.h	(revision 2764)
+++ source/Lib/TLibDecoder/TDecTop.h	(working copy)
@@ -116,7 +116,7 @@
   
   Void  deletePicBuffer();
 
-  Void executeDeblockAndAlf(UInt& ruiPOC, TComList<TComPic*>*& rpcListPic, Int& iSkipFrame,  Int& iPOCLastDisplay);
+  TComList<TComPic*> *executeDeblockAndAlf(UInt& ruiPOC, Int& iSkipFrame,  Int& iPOCLastDisplay);
 
 protected:
   Void  xGetNewPicBuffer  (TComSlice* pcSlice, TComPic*& rpcPic);
Index: source/Lib/TLibDecoder/TDecSbac.h
===================================================================
--- source/Lib/TLibDecoder/TDecSbac.h	(revision 2764)
+++ source/Lib/TLibDecoder/TDecSbac.h	(working copy)
@@ -85,7 +85,7 @@
 #endif
   void parseSEI(SEImessages&) {}
 
-  Void  parseSliceHeader          ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager) {}
+  Void  parseSliceHeader          ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager) {}
   Void  parseTerminatingBit       ( UInt& ruiBit );
   Void  parseMVPIdx               ( Int& riMVPIdx          );
   Void  parseSaoMaxUvlc           ( UInt& val, UInt maxSymbol );
Index: source/Lib/TLibDecoder/TDecGop.cpp
===================================================================
--- source/Lib/TLibDecoder/TDecGop.cpp	(revision 2764)
+++ source/Lib/TLibDecoder/TDecGop.cpp	(working copy)
@@ -109,7 +109,7 @@
 // Public member functions
 // ====================================================================================================================
 
-Void TDecGop::decompressSlice(TComInputBitstream* pcBitstream, TComPic*& rpcPic)
+Void TDecGop::decompressSlice(TComInputBitstream* pcBitstream, TComPic* rpcPic)
 {
   TComSlice*  pcSlice = rpcPic->getSlice(rpcPic->getCurrSliceIdx());
   // Table of extracted substreams.
@@ -198,7 +198,7 @@
   m_dDecTime += (double)(clock()-iBeforeTime) / CLOCKS_PER_SEC;
 }
 
-Void TDecGop::filterPicture(TComPic*& rpcPic)
+Void TDecGop::filterPicture(TComPic* rpcPic)
 {
   TComSlice*  pcSlice = rpcPic->getSlice(rpcPic->getCurrSliceIdx());
 
Index: source/Lib/TLibDecoder/TDecCAVLC.cpp
===================================================================
--- source/Lib/TLibDecoder/TDecCAVLC.cpp	(revision 2764)
+++ source/Lib/TLibDecoder/TDecCAVLC.cpp	(working copy)
@@ -802,7 +802,7 @@
   return;
 }
 
-Void TDecCavlc::parseSliceHeader (TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager)
+Void TDecCavlc::parseSliceHeader (TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager)
 {
   UInt  uiCode;
   Int   iCode;
@@ -1830,7 +1830,7 @@
       RefPicList  eRefPicList = ( iNumRef ? REF_PIC_LIST_1 : REF_PIC_LIST_0 );
       for ( Int iRefIdx=0 ; iRefIdx<pcSlice->getNumRefIdx(eRefPicList) ; iRefIdx++ ) 
       {
-        pcSlice->getWpScaling(eRefPicList, iRefIdx, wp);
+        wp = pcSlice->getWpScaling(eRefPicList, iRefIdx);
 
         wp[0].uiLog2WeightDenom = uiLog2WeightDenomLuma;
         wp[1].uiLog2WeightDenom = uiLog2WeightDenomChroma;
@@ -1847,7 +1847,7 @@
         UInt  uiCode;
         for ( Int iRefIdx=0 ; iRefIdx<pcSlice->getNumRefIdx(eRefPicList) ; iRefIdx++ ) 
         {
-          pcSlice->getWpScaling(eRefPicList, iRefIdx, wp);
+          wp = pcSlice->getWpScaling(eRefPicList, iRefIdx);
           READ_FLAG( uiCode, "chroma_weight_lX_flag" );      // u(1): chroma_weight_l0_flag
           wp[1].bPresentFlag = ( uiCode == 1 );
           wp[2].bPresentFlag = ( uiCode == 1 );
@@ -1856,7 +1856,7 @@
       }
       for ( Int iRefIdx=0 ; iRefIdx<pcSlice->getNumRefIdx(eRefPicList) ; iRefIdx++ ) 
       {
-        pcSlice->getWpScaling(eRefPicList, iRefIdx, wp);
+        wp = pcSlice->getWpScaling(eRefPicList, iRefIdx);
 #endif
         if ( wp[0].bPresentFlag ) 
         {
@@ -1909,7 +1909,7 @@
 
       for ( Int iRefIdx=pcSlice->getNumRefIdx(eRefPicList) ; iRefIdx<MAX_NUM_REF ; iRefIdx++ ) 
       {
-        pcSlice->getWpScaling(eRefPicList, iRefIdx, wp);
+        wp = pcSlice->getWpScaling(eRefPicList, iRefIdx);
 
         wp[0].bPresentFlag = false;
         wp[1].bPresentFlag = false;
Index: source/Lib/TLibDecoder/TDecGop.h
===================================================================
--- source/Lib/TLibDecoder/TDecGop.h	(revision 2764)
+++ source/Lib/TLibDecoder/TDecGop.h	(working copy)
@@ -113,8 +113,8 @@
   Void  create  ();
   Void  destroy ();
 //  Void  decompressGop(TComInputBitstream* pcBitstream, TComPic*& rpcPic, Bool bExecuteDeblockAndAlf );
-  Void  decompressSlice(TComInputBitstream* pcBitstream, TComPic*& rpcPic );
-  Void  filterPicture  (TComPic*& rpcPic );
+  Void  decompressSlice(TComInputBitstream* pcBitstream, TComPic* rpcPic );
+  Void  filterPicture  (TComPic* rpcPic );
   Void  setGopSize( Int i) { m_iGopSize = i; }
 
   void setPictureDigestEnabled(Int enabled) { m_pictureDigestEnabled = enabled; }
Index: source/Lib/TLibDecoder/TDecCAVLC.h
===================================================================
--- source/Lib/TLibDecoder/TDecCAVLC.h	(revision 2764)
+++ source/Lib/TLibDecoder/TDecCAVLC.h	(working copy)
@@ -108,7 +108,7 @@
 #if !REMOVE_APS
   Void  parseAPS            ( TComAPS* pAPS );
 #endif
-  Void  parseSliceHeader    ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager);
+  Void  parseSliceHeader    ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager);
   Void  parseTerminatingBit ( UInt& ruiBit );
   
   Void  parseMVPIdx         ( Int& riMVPIdx );
Index: source/Lib/TLibCommon/TComPic.h
===================================================================
--- source/Lib/TLibCommon/TComPic.h	(revision 2764)
+++ source/Lib/TLibCommon/TComPic.h	(working copy)
@@ -106,7 +106,7 @@
   TComPicSym*   getPicSym()           { return  m_apcPicSym;    }
   TComSlice*    getSlice(Int i)       { return  m_apcPicSym->getSlice(i);  }
   Int           getPOC()              { return  m_apcPicSym->getSlice(m_uiCurrSliceIdx)->getPOC();  }
-  TComDataCU*&  getCU( UInt uiCUAddr )  { return  m_apcPicSym->getCU( uiCUAddr ); }
+  TComDataCU*   getCU( UInt uiCUAddr )  { return  m_apcPicSym->getCU( uiCUAddr ); }
   
   TComPicYuv*   getPicYuvOrg()        { return  m_apcPicYuv[0]; }
   TComPicYuv*   getPicYuvRec()        { return  m_apcPicYuv[1]; }
Index: source/Lib/TLibCommon/TComMotionInfo.h
===================================================================
--- source/Lib/TLibCommon/TComMotionInfo.h	(revision 2764)
+++ source/Lib/TLibCommon/TComMotionInfo.h	(working copy)
@@ -93,7 +93,7 @@
 private:
   TComMv*   m_pcMv;
   TComMv*   m_pcMvd;
-  Char*     m_piRefIdx;
+  Short*    m_piRefIdx;
   UInt      m_uiNumPartition;
   AMVPInfo  m_cAMVPInfo;
     
@@ -152,7 +152,7 @@
     m_piRefIdx = src->m_piRefIdx + offset;
   }
   
-  Void compress(Char* pePredMode, Int scale); 
+  Void compress(UChar* pePredMode, Int scale); 
 };
 
 //! \}
Index: source/Lib/TLibCommon/TComCABACTables.h
===================================================================
--- source/Lib/TLibCommon/TComCABACTables.h	(revision 2764)
+++ source/Lib/TLibCommon/TComCABACTables.h	(working copy)
@@ -47,14 +47,9 @@
  * \brief static class for CABAC tables
  */
 
-class TComCABACTables
-{
-public:
-  const static UChar  sm_aucLPSTable[64][4];
-  const static UChar  sm_aucRenormTable[32];
-};
+extern const UChar sm_aucLPSTable[64][4];
+extern const UChar sm_aucRenormTable[32];
 
-
 //! \}
 
 #endif
Index: source/Lib/TLibCommon/NAL.h
===================================================================
--- source/Lib/TLibCommon/NAL.h	(revision 2764)
+++ source/Lib/TLibCommon/NAL.h	(working copy)
@@ -88,7 +88,8 @@
  */
 struct NALUnitEBSP : public NALUnit
 {
-  std::ostringstream m_nalUnitData;
+//std::ostringstream m_nalUnitData;
+  std::ostream *m_nalUnitData;
 
   /**
    * convert the OutputNALUnit #nalu# into EBSP format by writing out
Index: source/Lib/TLibCommon/TComRom.h
===================================================================
--- source/Lib/TLibCommon/TComRom.h	(revision 2764)
+++ source/Lib/TLibCommon/TComRom.h	(working copy)
@@ -53,7 +53,8 @@
 #define     MAX_CU_DEPTH            7                           // log2(LCUSize)
 #define     MAX_CU_SIZE             (1<<(MAX_CU_DEPTH))         // maximum allowable size of CU
 #define     MIN_PU_SIZE             4
-#define     MAX_NUM_SPU_W           (MAX_CU_SIZE/MIN_PU_SIZE)   // maximum number of SPU in horizontal line
+//#define     MAX_NUM_SPU_W           (MAX_CU_SIZE/MIN_PU_SIZE)   // maximum number of SPU in horizontal line
+#define     MAX_NUM_SPU_W           (32)   // maximum number of SPU in horizontal line
 
 // ====================================================================================================================
 // Initialize / destroy functions
@@ -75,7 +76,7 @@
 extern       UInt   g_auiRasterToZscan[ MAX_NUM_SPU_W*MAX_NUM_SPU_W ];
 extern       UInt   g_motionRefer[ MAX_NUM_SPU_W*MAX_NUM_SPU_W ];
 
-Void         initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt*& rpuiCurrIdx );
+UInt        *initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt* rpuiCurrIdx );
 Void         initRasterToZscan ( UInt uiMaxCUWidth, UInt uiMaxCUHeight, UInt uiMaxDepth         );
 
 Void          initMotionReferIdx ( UInt uiMaxCUWidth, UInt uiMaxCUHeight, UInt uiMaxDepth );
Index: source/Lib/TLibCommon/TComSlice.h
===================================================================
--- source/Lib/TLibCommon/TComSlice.h	(revision 2764)
+++ source/Lib/TLibCommon/TComSlice.h	(working copy)
@@ -496,7 +496,7 @@
   Void setScalingListFlag       ( Bool b ) { m_scalingListEnabledFlag  = b;       }
   Bool getScalingListPresentFlag()         { return m_scalingListPresentFlag;     }
   Void setScalingListPresentFlag( Bool b ) { m_scalingListPresentFlag  = b;       }
-  Void setScalingList      ( TComScalingList *scalingList);
+  Void setScalingList      ( TComScalingList *scalingList) {}
   TComScalingList* getScalingList ()       { return m_scalingList; }               //!< get ScalingList class pointer in SPS
   UInt getMaxDecPicBuffering  (UInt tlayer)            { return m_uiMaxDecPicBuffering[tlayer]; }
   Void setMaxDecPicBuffering  ( UInt ui, UInt tlayer ) { m_uiMaxDecPicBuffering[tlayer] = ui;   }
@@ -517,8 +517,8 @@
   TComRefPicListModification();
   virtual ~TComRefPicListModification();
   
-  Void  create                    ();
-  Void  destroy                   ();
+  Void  create                    () {}
+  Void  destroy                   () {}
 
   Bool       getRefPicListModificationFlagL0() { return m_bRefPicListModificationFlagL0; }
   Void       setRefPicListModificationFlagL0(Bool flag) { m_bRefPicListModificationFlagL0 = flag; }
@@ -718,7 +718,7 @@
   Int      getLoopFilterTcOffset()             {return m_loopFilterTcOffsetDiv2; }     //!< get tc offset for deblocking filter
   Bool     getScalingListPresentFlag()         { return m_scalingListPresentFlag;     }
   Void     setScalingListPresentFlag( Bool b ) { m_scalingListPresentFlag  = b;       }
-  Void     setScalingList      ( TComScalingList *scalingList);
+  Void     setScalingList      ( TComScalingList *scalingList) {}
   TComScalingList* getScalingList ()          { return m_scalingList; }         //!< get ScalingList class pointer in PPS
   UInt getLog2ParallelMergeLevelMinus2      ()                    { return m_log2ParallelMergeLevelMinus2; }
   Void setLog2ParallelMergeLevelMinus2      (UInt mrgLevel)       { m_log2ParallelMergeLevelMinus2 = mrgLevel; }
@@ -1106,7 +1106,7 @@
   Void setTLayer             ( UInt uiTLayer )             { m_uiTLayer = uiTLayer;                  }
 
   Void setTLayerInfo( UInt uiTLayer );
-  Void decodingMarking( TComList<TComPic*>& rcListPic, Int iGOPSIze, Int& iMaxRefPicNum ); 
+  Void decodingMarking( TComList<TComPic*>& rcListPic, Int iGOPSIze, Int& iMaxRefPicNum ) {}
   Void applyReferencePictureSet( TComList<TComPic*>& rcListPic, TComReferencePictureSet *RPSList);
   Bool isTemporalLayerSwitchingPoint( TComList<TComPic*>& rcListPic, TComReferencePictureSet *RPSList);
   Int       checkThatAllRefPicsAreAvailable( TComList<TComPic*>& rcListPic, TComReferencePictureSet *pReferencePictureSet, Bool printErrors, Int pocRandomAccess = 0);
@@ -1146,7 +1146,7 @@
   Void setFinalized                     ( Bool uiVal )      { m_bFinalized = uiVal;                       }
   Bool getFinalized                     ()                  { return m_bFinalized;                        }
   Void  setWpScaling    ( wpScalingParam  wp[2][MAX_NUM_REF][3] ) { memcpy(m_weightPredTable, wp, sizeof(wpScalingParam)*2*MAX_NUM_REF*3); }
-  Void  getWpScaling    ( RefPicList e, Int iRefIdx, wpScalingParam *&wp);
+  wpScalingParam* getWpScaling    ( RefPicList e, Int iRefIdx);
 
   Void  resetWpScaling  (wpScalingParam  wp[2][MAX_NUM_REF][3]);
   Void  initWpScaling    (wpScalingParam  wp[2][MAX_NUM_REF][3]);
Index: source/Lib/TLibCommon/TComPrediction.h
===================================================================
--- source/Lib/TLibCommon/TComPrediction.h	(revision 2764)
+++ source/Lib/TLibCommon/TComPrediction.h	(working copy)
@@ -98,7 +98,7 @@
   Void motionCompensation         ( TComDataCU*  pcCU, TComYuv* pcYuvPred, RefPicList eRefPicList = REF_PIC_LIST_X, Int iPartIdx = -1 );
   
   // motion vector prediction
-  Void getMvPredAMVP              ( TComDataCU* pcCU, UInt uiPartIdx, UInt uiPartAddr, RefPicList eRefPicList, Int iRefIdx, TComMv& rcMvPred );
+  TComMv getMvPredAMVP              ( TComDataCU* pcCU, UInt uiPartIdx, UInt uiPartAddr, RefPicList eRefPicList, Int iRefIdx);
   
   // Angular Intra
   Void predIntraLumaAng           ( TComPattern* pcTComPattern, UInt uiDirMode, Pel* piPred, UInt uiStride, Int iWidth, Int iHeight,  TComDataCU* pcCU, Bool bAbove, Bool bLeft );
Index: source/Lib/TLibCommon/TComDataCU.h
===================================================================
--- source/Lib/TLibCommon/TComDataCU.h	(revision 2764)
+++ source/Lib/TLibCommon/TComDataCU.h	(working copy)
@@ -130,10 +130,10 @@
 #if SKIP_FLAG
   Bool*         m_skipFlag;           ///< array of skip flags
 #endif
-  Char*         m_pePartSize;         ///< array of partition sizes
-  Char*         m_pePredMode;         ///< array of prediction modes
+  UChar*        m_pePartSize;         ///< array of partition sizes
+  UChar*        m_pePredMode;         ///< array of prediction modes
   Bool*         m_CUTransquantBypass;   ///< array of cu_transquant_bypass flags
-  Char*         m_phQP;               ///< array of QP values
+  UChar*        m_phQP;               ///< array of QP values
   UChar*        m_puhTrIdx;           ///< array of transform indices
   UChar*        m_puhTransformSkip[3];///< array of transform skipping flags
 #if !REMOVE_NSQT
@@ -188,8 +188,8 @@
   UChar*        m_puhLumaIntraDir;    ///< array of intra directions (luma)
   UChar*        m_puhChromaIntraDir;  ///< array of intra directions (chroma)
   UChar*        m_puhInterDir;        ///< array of inter directions
-  Char*         m_apiMVPIdx[2];       ///< array of motion vector predictor candidates
-  Char*         m_apiMVPNum[2];       ///< array of number of possible motion vectors predictors
+  UChar*        m_apiMVPIdx[2];       ///< array of motion vector predictor candidates
+  UChar*        m_apiMVPNum[2];       ///< array of number of possible motion vectors predictors
   Bool          m_lcuAlfEnabled[3];
   Bool*         m_pbIPCMFlag;         ///< array of intra_pcm flags
 
@@ -207,7 +207,7 @@
   UInt          m_uiTotalBins;       ///< sum of partition bins
   UInt*         m_uiSliceStartCU;    ///< Start CU address of current slice
   UInt*         m_uiDependentSliceStartCU; ///< Start CU address of current slice
-  Char          m_codedQP;
+  UChar         m_codedQP;
 protected:
   
   /// add possible motion vector predictor candidates
@@ -263,8 +263,8 @@
   
   TComPic*      getPic                ()                        { return m_pcPic;           }
   TComSlice*    getSlice              ()                        { return m_pcSlice;         }
-  UInt&         getAddr               ()                        { return m_uiCUAddr;        }
-  UInt&         getZorderIdxInCU      ()                        { return m_uiAbsIdxInLCU; }
+  UInt          getAddr               ()                        { return m_uiCUAddr;        }
+  UInt          getZorderIdxInCU      ()                        { return m_uiAbsIdxInLCU; }
   UInt          getSCUAddr            ();
   UInt          getCUPelX             ()                        { return m_uiCUPelX;        }
   UInt          getCUPelY             ()                        { return m_uiCUPelY;        }
@@ -280,7 +280,7 @@
   // member functions for CU data
   // -------------------------------------------------------------------------------------------------------------------
   
-  Char*         getPartitionSize      ()                        { return m_pePartSize;        }
+  UChar*        getPartitionSize      ()                        { return m_pePartSize;        }
   PartSize      getPartitionSize      ( UInt uiIdx )            { return static_cast<PartSize>( m_pePartSize[uiIdx] ); }
   Void          setPartitionSize      ( UInt uiIdx, PartSize uh){ m_pePartSize[uiIdx] = uh;   }
   Void          setPartSizeSubParts   ( PartSize eMode, UInt uiAbsPartIdx, UInt uiDepth );
@@ -293,7 +293,7 @@
   Void         setSkipFlagSubParts   ( Bool skip, UInt absPartIdx, UInt depth );
 #endif
 
-  Char*         getPredictionMode     ()                        { return m_pePredMode;        }
+  UChar*        getPredictionMode     ()                        { return m_pePredMode;        }
   PredMode      getPredictionMode     ( UInt uiIdx )            { return static_cast<PredMode>( m_pePredMode[uiIdx] ); }
   Bool*         getCUTransquantBypass ()                        { return m_CUTransquantBypass;        }
   Bool          getCUTransquantBypass( UInt uiIdx )             { return m_CUTransquantBypass[uiIdx]; }
@@ -310,15 +310,15 @@
   
   Void          setSizeSubParts       ( UInt uiWidth, UInt uiHeight, UInt uiAbsPartIdx, UInt uiDepth );
   
-  Char*         getQP                 ()                        { return m_phQP;              }
-  Char          getQP                 ( UInt uiIdx )            { return m_phQP[uiIdx];       }
-  Void          setQP                 ( UInt uiIdx, Char value ){ m_phQP[uiIdx] =  value;     }
+  UChar*        getQP                 ()                        { return m_phQP;              }
+  UChar         getQP                 ( UInt uiIdx )            { return m_phQP[uiIdx];       }
+  Void          setQP                 ( UInt uiIdx, Short value ){ m_phQP[uiIdx] =  value;     }
   Void          setQPSubParts         ( Int qp,   UInt uiAbsPartIdx, UInt uiDepth );
   Int           getLastValidPartIdx   ( Int iAbsPartIdx );
-  Char          getLastCodedQP        ( UInt uiAbsPartIdx );
+  UChar         getLastCodedQP        ( UInt uiAbsPartIdx );
   Void          setQPSubCUs           ( Int qp, TComDataCU* pcCU, UInt absPartIdx, UInt depth, Bool &foundNonZeroCbf );
-  Void          setCodedQP            ( Char qp )               { m_codedQP = qp;             }
-  Char          getCodedQP            ()                        { return m_codedQP;           }
+  Void          setCodedQP            ( Short qp )              { m_codedQP = qp;             }
+  UChar         getCodedQP            ()                        { return m_codedQP;           }
 
   Bool          isLosslessCoded(UInt absPartIdx);
 #if !REMOVE_NSQT
@@ -341,18 +341,18 @@
   
   TComCUMvField* getCUMvField         ( RefPicList e )          { return  &m_acCUMvField[e];  }
   
-  TCoeff*&      getCoeffY             ()                        { return m_pcTrCoeffY;        }
-  TCoeff*&      getCoeffCb            ()                        { return m_pcTrCoeffCb;       }
-  TCoeff*&      getCoeffCr            ()                        { return m_pcTrCoeffCr;       }
+  TCoeff*       getCoeffY             ()                        { return m_pcTrCoeffY;        }
+  TCoeff*       getCoeffCb            ()                        { return m_pcTrCoeffCb;       }
+  TCoeff*       getCoeffCr            ()                        { return m_pcTrCoeffCr;       }
 #if ADAPTIVE_QP_SELECTION
-  Int*&         getArlCoeffY          ()                        { return m_pcArlCoeffY;       }
-  Int*&         getArlCoeffCb         ()                        { return m_pcArlCoeffCb;      }
-  Int*&         getArlCoeffCr         ()                        { return m_pcArlCoeffCr;      }
+  Int*          getArlCoeffY          ()                        { return m_pcArlCoeffY;       }
+  Int*          getArlCoeffCb         ()                        { return m_pcArlCoeffCb;      }
+  Int*          getArlCoeffCr         ()                        { return m_pcArlCoeffCr;      }
 #endif
   
-  Pel*&         getPCMSampleY         ()                        { return m_pcIPCMSampleY;     }
-  Pel*&         getPCMSampleCb        ()                        { return m_pcIPCMSampleCb;    }
-  Pel*&         getPCMSampleCr        ()                        { return m_pcIPCMSampleCr;    }
+  Pel*          getPCMSampleY         ()                        { return m_pcIPCMSampleY;     }
+  Pel*          getPCMSampleCb        ()                        { return m_pcIPCMSampleCb;    }
+  Pel*          getPCMSampleCr        ()                        { return m_pcIPCMSampleCr;    }
 
   UChar         getCbf    ( UInt uiIdx, TextType eType )                  { return m_puhCbf[g_aucConvertTxtTypeToIdx[eType]][uiIdx];  }
   UChar*        getCbf    ( TextType eType )                              { return m_puhCbf[g_aucConvertTxtTypeToIdx[eType]];         }
@@ -446,11 +446,11 @@
   Void          getPartPosition       ( UInt partIdx, Int& xP, Int& yP, Int& nPSW, Int& nPSH);
   Void          setMVPIdx             ( RefPicList eRefPicList, UInt uiIdx, Int iMVPIdx)  { m_apiMVPIdx[eRefPicList][uiIdx] = iMVPIdx;  }
   Int           getMVPIdx             ( RefPicList eRefPicList, UInt uiIdx)               { return m_apiMVPIdx[eRefPicList][uiIdx];     }
-  Char*         getMVPIdx             ( RefPicList eRefPicList )                          { return m_apiMVPIdx[eRefPicList];            }
+  UChar*        getMVPIdx             ( RefPicList eRefPicList )                          { return m_apiMVPIdx[eRefPicList];            }
 
   Void          setMVPNum             ( RefPicList eRefPicList, UInt uiIdx, Int iMVPNum ) { m_apiMVPNum[eRefPicList][uiIdx] = iMVPNum;  }
   Int           getMVPNum             ( RefPicList eRefPicList, UInt uiIdx )              { return m_apiMVPNum[eRefPicList][uiIdx];     }
-  Char*         getMVPNum             ( RefPicList eRefPicList )                          { return m_apiMVPNum[eRefPicList];            }
+  UChar*        getMVPNum             ( RefPicList eRefPicList )                          { return m_apiMVPNum[eRefPicList];            }
   
   Void          setMVPIdxSubParts     ( Int iMVPIdx, RefPicList eRefPicList, UInt uiAbsPartIdx, UInt uiPartIdx, UInt uiDepth );
   Void          setMVPNumSubParts     ( Int iMVPNum, RefPicList eRefPicList, UInt uiAbsPartIdx, UInt uiPartIdx, UInt uiDepth );
@@ -493,7 +493,7 @@
 
   TComDataCU*   getQpMinCuLeft              ( UInt&  uiLPartUnitIdx , UInt uiCurrAbsIdxInLCU, Bool bEnforceSliceRestriction=true, Bool bEnforceDependentSliceRestriction=true );
   TComDataCU*   getQpMinCuAbove             ( UInt&  aPartUnitIdx , UInt currAbsIdxInLCU, Bool enforceSliceRestriction=true, Bool enforceDependentSliceRestriction=true );
-  Char          getRefQP                    ( UInt   uiCurrAbsIdxInLCU                       );
+  UChar         getRefQP                    ( UInt   uiCurrAbsIdxInLCU                       );
 
   TComDataCU*   getPUAboveRightAdi          ( UInt&  uiARPartUnitIdx, UInt uiPuWidth, UInt uiCurrPartUnitIdx, UInt uiPartUnitOffset = 1, Bool bEnforceSliceRestriction=true, Bool bEnforceDependentSliceRestriction=true );
   TComDataCU*   getPUBelowLeftAdi           ( UInt&  uiBLPartUnitIdx, UInt uiPuHeight, UInt uiCurrPartUnitIdx, UInt uiPartUnitOffset = 1, Bool bEnforceSliceRestriction=true, Bool bEnforceDependentSliceRestriction=true );
@@ -548,7 +548,7 @@
   Double&       getTotalCost()                  { return m_dTotalCost;        }
   UInt&         getTotalDistortion()            { return m_uiTotalDistortion; }
   UInt&         getTotalBits()                  { return m_uiTotalBits;       }
-  UInt&         getTotalNumPart()               { return m_uiNumPartition;    }
+  UInt          getTotalNumPart()               { return m_uiNumPartition;    }
 
   UInt          getCoefScanIdx(UInt uiAbsPartIdx, UInt uiWidth, Bool bIsLuma, Bool bIsIntra);
 
Index: source/Lib/TLibCommon/TComWeightPrediction.cpp
===================================================================
--- source/Lib/TLibCommon/TComWeightPrediction.cpp	(revision 2764)
+++ source/Lib/TLibCommon/TComWeightPrediction.cpp	(working copy)
@@ -262,11 +262,11 @@
   { // explicit --------------------
     if ( iRefIdx0 >= 0 )
     {
-      pcSlice->getWpScaling(REF_PIC_LIST_0, iRefIdx0, wp0);
+      wp0 = pcSlice->getWpScaling(REF_PIC_LIST_0, iRefIdx0);
     }
     if ( iRefIdx1 >= 0 )
     {
-      pcSlice->getWpScaling(REF_PIC_LIST_1, iRefIdx1, wp1);
+      wp1 = pcSlice->getWpScaling(REF_PIC_LIST_1, iRefIdx1);
     }
   }
   else
Index: source/Lib/TLibCommon/TComList.h
===================================================================
--- source/Lib/TLibCommon/TComList.h	(revision 2764)
+++ source/Lib/TLibCommon/TComList.h	(working copy)
@@ -67,7 +67,7 @@
   {
     if( ! rcTComList.empty() )
     {
-      insert( this->end(), rcTComList.begin(), rcTComList.end());
+      this->insert( this->end(), rcTComList.begin(), rcTComList.end());
     }
     return *this;
   } // leszek
@@ -104,10 +104,10 @@
     }
   }
   
-  TComIterator find( const C& rcT ) // leszek
+/*TComIterator find( const C& rcT ) // leszek
   {
-    return find( this->begin(), this->end(), rcT );
-  }
+    return this->find( this->begin(), this->end(), rcT );
+  }*/
 };
 
 //! \}
Index: source/Lib/TLibCommon/TComMotionInfo.cpp
===================================================================
--- source/Lib/TLibCommon/TComMotionInfo.cpp	(revision 2764)
+++ source/Lib/TLibCommon/TComMotionInfo.cpp	(working copy)
@@ -59,7 +59,7 @@
   
   m_pcMv     = new TComMv[ uiNumPartition ];
   m_pcMvd    = new TComMv[ uiNumPartition ];
-  m_piRefIdx = new Char  [ uiNumPartition ];
+  m_piRefIdx = new Short [ uiNumPartition ];
   
   m_uiNumPartition = uiNumPartition;
 }
@@ -92,7 +92,7 @@
     m_pcMv [ i ].setZero();
     m_pcMvd[ i ].setZero();      
   }
-  assert( sizeof( *m_piRefIdx ) == 1 );
+  assert( sizeof( *m_piRefIdx ) == 2 );
   memset( m_piRefIdx, NOT_VALID, m_uiNumPartition * sizeof( *m_piRefIdx ) );
 }
 
@@ -314,7 +314,7 @@
 
 Void TComCUMvField::setAllRefIdx ( Int iRefIdx, PartSize eCUMode, Int iPartAddr, UInt uiDepth, Int iPartIdx )
 {
-  setAll(m_piRefIdx, static_cast<Char>(iRefIdx), eCUMode, iPartAddr, uiDepth, iPartIdx);
+  setAll(m_piRefIdx, static_cast<Short>(iRefIdx), eCUMode, iPartAddr, uiDepth, iPartIdx);
 }
 
 Void TComCUMvField::setAllMvField( TComMvField const & mvField, PartSize eCUMode, Int iPartAddr, UInt uiDepth, Int iPartIdx )
@@ -327,7 +327,7 @@
  * \param pePredMode Pointer to prediction modes
  * \param scale      Factor by which to subsample motion information
  */
-Void TComCUMvField::compress(Char* pePredMode, Int scale)
+Void TComCUMvField::compress(UChar* pePredMode, Int scale)
 {
   Int N = scale * scale;
   assert( N > 0 && N <= m_uiNumPartition);
Index: source/Lib/TLibCommon/TComCABACTables.cpp
===================================================================
--- source/Lib/TLibCommon/TComCABACTables.cpp	(revision 2764)
+++ source/Lib/TLibCommon/TComCABACTables.cpp	(working copy)
@@ -40,7 +40,7 @@
 //! \ingroup TLibCommon
 //! \{
 
-const UChar TComCABACTables::sm_aucLPSTable[64][4] =
+const UChar sm_aucLPSTable[64][4] =
 {
   { 128, 176, 208, 240},
   { 128, 167, 197, 227},
@@ -108,7 +108,7 @@
   {   2,   2,   2,   2}
 };
 
-const UChar TComCABACTables::sm_aucRenormTable[32] =
+const UChar sm_aucRenormTable[32] =
 {
   6,  5,  4,  4,
   3,  3,  3,  3,
Index: source/Lib/TLibCommon/TComRom.cpp
===================================================================
--- source/Lib/TLibCommon/TComRom.cpp	(revision 2764)
+++ source/Lib/TLibCommon/TComRom.cpp	(working copy)
@@ -179,7 +179,7 @@
 
 UInt g_auiPUOffset[8] = { 0, 8, 4, 4, 2, 10, 1, 5};
 
-Void initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt*& rpuiCurrIdx )
+UInt *initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt* rpuiCurrIdx )
 {
   Int iStride = 1 << ( iMaxDepth - 1 );
   
@@ -191,11 +191,13 @@
   else
   {
     Int iStep = iStride >> iDepth;
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal,                     rpuiCurrIdx );
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep,               rpuiCurrIdx );
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride,       rpuiCurrIdx );
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride+iStep, rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal,                     rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep,               rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride,       rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride+iStep, rpuiCurrIdx );
   }
+
+  return rpuiCurrIdx;
 }
 
 Void initRasterToZscan ( UInt uiMaxCUWidth, UInt uiMaxCUHeight, UInt uiMaxDepth )
@@ -402,6 +404,8 @@
   {84,  -29,   -74,   55},
   {55,  -84,    74,  -29},
 };
+const UChar g_aucDCTDSTMode_Vert[NUM_INTRA_MODE] = {};
+const UChar g_aucDCTDSTMode_Hor[NUM_INTRA_MODE] = {};
 
 
 // ====================================================================================================================
@@ -432,6 +436,10 @@
 };
 #endif // FAST_UDI_USE_MPM
 
+const UChar g_aucIntraModeNumAng[7] = {};
+const UChar g_aucIntraModeBitsAng[7] = {};
+const UChar g_aucAngIntraModeOrder[NUM_INTRA_MODE] = {};
+
 // chroma
 
 const UChar g_aucConvertTxtTypeToIdx[4] = { 0, 1, 1, 2 };
@@ -826,6 +834,10 @@
   20,24,25,28,33,41,54,71,
   24,25,28,33,41,54,71,91
 };
+Int g_quantIntraDefault16x16[256] = {};
+Int g_quantIntraDefault32x32[1024] = {};
+Int g_quantInterDefault16x16[256] = {};
+Int g_quantInterDefault32x32[1024] = {};
 UInt g_scalingListSize   [4] = {16,64,256,1024}; 
 UInt g_scalingListSizeX  [4] = { 4, 8, 16,  32};
 UInt g_scalingListNum[SCALING_LIST_SIZE_NUM]={6,6,6,2};
Index: source/Lib/TLibCommon/TComSlice.cpp
===================================================================
--- source/Lib/TLibCommon/TComSlice.cpp	(revision 2764)
+++ source/Lib/TLibCommon/TComSlice.cpp	(working copy)
@@ -1161,9 +1161,9 @@
  * \param *&wpScalingParam
  * \returns Void
  */
-Void  TComSlice::getWpScaling( RefPicList e, Int iRefIdx, wpScalingParam *&wp )
+wpScalingParam* TComSlice::getWpScaling( RefPicList e, Int iRefIdx)
 {
-  wp = m_weightPredTable[e][iRefIdx];
+  return m_weightPredTable[e][iRefIdx];
 }
 
 /** reset Default WP tables settings : no weight. 
Index: source/Lib/TLibCommon/TComPrediction.cpp
===================================================================
--- source/Lib/TLibCommon/TComPrediction.cpp	(revision 2764)
+++ source/Lib/TLibCommon/TComPrediction.cpp	(working copy)
@@ -658,8 +658,9 @@
 }
 
 // AMVP
-Void TComPrediction::getMvPredAMVP( TComDataCU* pcCU, UInt uiPartIdx, UInt uiPartAddr, RefPicList eRefPicList, Int iRefIdx, TComMv& rcMvPred )
+TComMv TComPrediction::getMvPredAMVP( TComDataCU* pcCU, UInt uiPartIdx, UInt uiPartAddr, RefPicList eRefPicList, Int iRefIdx)
 {
+  TComMv rcMvPred;
   AMVPInfo* pcAMVPInfo = pcCU->getCUMvField(eRefPicList)->getAMVPInfo();
 
   if( pcCU->getAMVPMode(uiPartAddr) == AM_NONE || (pcAMVPInfo->iN <= 1 && pcCU->getAMVPMode(uiPartAddr) == AM_EXPL) )
@@ -668,12 +669,12 @@
 
     pcCU->setMVPIdxSubParts( 0, eRefPicList, uiPartAddr, uiPartIdx, pcCU->getDepth(uiPartAddr));
     pcCU->setMVPNumSubParts( pcAMVPInfo->iN, eRefPicList, uiPartAddr, uiPartIdx, pcCU->getDepth(uiPartAddr));
-    return;
+    return rcMvPred;
   }
 
-  assert(pcCU->getMVPIdx(eRefPicList,uiPartAddr) >= 0);
+  assert(pcCU->getMVPIdx(eRefPicList,uiPartAddr) != 255);
   rcMvPred = pcAMVPInfo->m_acMvCand[pcCU->getMVPIdx(eRefPicList,uiPartAddr)];
-  return;
+  return rcMvPred;
 }
 
 /** Function for deriving planar intra prediction.
Index: source/Lib/TLibCommon/TComDataCU.cpp
===================================================================
--- source/Lib/TLibCommon/TComDataCU.cpp	(revision 2764)
+++ source/Lib/TLibCommon/TComDataCU.cpp	(working copy)
@@ -140,7 +140,7 @@
   
   if ( !bDecSubCu )
   {
-    m_phQP               = (Char*     )xMalloc(Char,     uiNumPartition);
+    m_phQP               = (UChar*    )xMalloc(UChar,    uiNumPartition);
     m_puhDepth           = (UChar*    )xMalloc(UChar,    uiNumPartition);
     m_puhWidth           = (UChar*    )xMalloc(UChar,    uiNumPartition);
     m_puhHeight          = (UChar*    )xMalloc(UChar,    uiNumPartition);
@@ -149,9 +149,9 @@
     m_skipFlag           = new Bool[ uiNumPartition ];
 #endif
 
-    m_pePartSize         = new Char[ uiNumPartition ];
+    m_pePartSize         = new UChar[ uiNumPartition ];
     memset( m_pePartSize, SIZE_NONE,uiNumPartition * sizeof( *m_pePartSize ) );
-    m_pePredMode         = new Char[ uiNumPartition ];
+    m_pePredMode         = new UChar[ uiNumPartition ];
     m_CUTransquantBypass = new Bool[ uiNumPartition ];
     m_pbMergeFlag        = (Bool*  )xMalloc(Bool,   uiNumPartition);
     m_puhMergeIndex      = (UChar* )xMalloc(UChar,  uiNumPartition);
@@ -171,12 +171,12 @@
     m_puhCbf[1]          = (UChar* )xMalloc(UChar,  uiNumPartition);
     m_puhCbf[2]          = (UChar* )xMalloc(UChar,  uiNumPartition);
     
-    m_apiMVPIdx[0]       = new Char[ uiNumPartition ];
-    m_apiMVPIdx[1]       = new Char[ uiNumPartition ];
-    m_apiMVPNum[0]       = new Char[ uiNumPartition ];
-    m_apiMVPNum[1]       = new Char[ uiNumPartition ];
-    memset( m_apiMVPIdx[0], -1,uiNumPartition * sizeof( Char ) );
-    memset( m_apiMVPIdx[1], -1,uiNumPartition * sizeof( Char ) );
+    m_apiMVPIdx[0]       = new UChar[ uiNumPartition ];
+    m_apiMVPIdx[1]       = new UChar[ uiNumPartition ];
+    m_apiMVPNum[0]       = new UChar[ uiNumPartition ];
+    m_apiMVPNum[1]       = new UChar[ uiNumPartition ];
+    memset( m_apiMVPIdx[0], 255,uiNumPartition * sizeof( UChar ) );
+    memset( m_apiMVPIdx[1], 255,uiNumPartition * sizeof( UChar ) );
     
     m_pcTrCoeffY         = (TCoeff*)xMalloc(TCoeff, uiWidth*uiHeight);
     m_pcTrCoeffCb        = (TCoeff*)xMalloc(TCoeff, uiWidth*uiHeight/4);
@@ -465,10 +465,10 @@
     memset( m_puhTransformSkip[2] + firstElement, 0,                      numElements * sizeof( *m_puhTransformSkip[2]) );
     memset( m_puhWidth          + firstElement, g_uiMaxCUWidth,           numElements * sizeof( *m_puhWidth ) );
     memset( m_puhHeight         + firstElement, g_uiMaxCUHeight,          numElements * sizeof( *m_puhHeight ) );
-    memset( m_apiMVPIdx[0]      + firstElement, -1,                       numElements * sizeof( *m_apiMVPIdx[0] ) );
-    memset( m_apiMVPIdx[1]      + firstElement, -1,                       numElements * sizeof( *m_apiMVPIdx[1] ) );
-    memset( m_apiMVPNum[0]      + firstElement, -1,                       numElements * sizeof( *m_apiMVPNum[0] ) );
-    memset( m_apiMVPNum[1]      + firstElement, -1,                       numElements * sizeof( *m_apiMVPNum[1] ) );
+    memset( m_apiMVPIdx[0]      + firstElement, 255,                      numElements * sizeof( *m_apiMVPIdx[0] ) );
+    memset( m_apiMVPIdx[1]      + firstElement, 255,                      numElements * sizeof( *m_apiMVPIdx[1] ) );
+    memset( m_apiMVPNum[0]      + firstElement, 255,                      numElements * sizeof( *m_apiMVPNum[0] ) );
+    memset( m_apiMVPNum[1]      + firstElement, 255,                      numElements * sizeof( *m_apiMVPNum[1] ) );
     memset( m_phQP              + firstElement, getSlice()->getSliceQp(), numElements * sizeof( *m_phQP ) );
     m_lcuAlfEnabled[0] = m_lcuAlfEnabled[1] = m_lcuAlfEnabled[2] = false;
     memset( m_pbMergeFlag       + firstElement, false,                    numElements * sizeof( *m_pbMergeFlag ) );
@@ -591,10 +591,10 @@
   {
     if(getPic()->getPicSym()->getInverseCUOrderMap(getAddr())*m_pcPic->getNumPartInCU()+m_uiAbsIdxInLCU+ui >= getSlice()->getDependentSliceCurStartCUAddr())
     {
-      m_apiMVPIdx[0][ui] = -1;
-      m_apiMVPIdx[1][ui] = -1;
-      m_apiMVPNum[0][ui] = -1;
-      m_apiMVPNum[1][ui] = -1;
+      m_apiMVPIdx[0][ui] = 255;
+      m_apiMVPIdx[1][ui] = 255;
+      m_apiMVPNum[0][ui] = 255;
+      m_apiMVPNum[1][ui] = 255;
       m_puhDepth  [ui] = uiDepth;
       m_puhWidth  [ui] = uhWidth;
       m_puhHeight [ui] = uhHeight;
@@ -711,10 +711,10 @@
     m_pePartSize[ui] = SIZE_NONE;
     m_pePredMode[ui] = MODE_NONE;
     m_CUTransquantBypass[ui] = false;
-    m_apiMVPIdx[0][ui] = -1;
-    m_apiMVPIdx[1][ui] = -1;
-    m_apiMVPNum[0][ui] = -1;
-    m_apiMVPNum[1][ui] = -1;
+    m_apiMVPIdx[0][ui] = 255;
+    m_apiMVPIdx[1][ui] = 255;
+    m_apiMVPNum[0][ui] = 255;
+    m_apiMVPNum[1][ui] = 255;
     if(m_pcPic->getPicSym()->getInverseCUOrderMap(getAddr())*m_pcPic->getNumPartInCU()+m_uiAbsIdxInLCU+ui<getSlice()->getDependentSliceCurStartCUAddr())
     {
       m_apiMVPIdx[0][ui] = pcCU->m_apiMVPIdx[0][uiPartOffset+ui];
@@ -1057,7 +1057,7 @@
 // It is used to predict for next part
 Void TComDataCU::copyToPic( UChar uhDepth )
 {
-  TComDataCU*& rpcCU = m_pcPic->getCU( m_uiCUAddr );
+  TComDataCU* rpcCU = m_pcPic->getCU( m_uiCUAddr );
   
   rpcCU->getTotalCost()       = m_dTotalCost;
   rpcCU->getTotalDistortion() = m_uiTotalDistortion;
@@ -1135,7 +1135,7 @@
 
 Void TComDataCU::copyToPic( UChar uhDepth, UInt uiPartIdx, UInt uiPartDepth )
 {
-  TComDataCU*&  rpcCU       = m_pcPic->getCU( m_uiCUAddr );
+  TComDataCU*   rpcCU       = m_pcPic->getCU( m_uiCUAddr );
   UInt          uiQNumPart  = m_uiNumPartition>>(uiPartDepth<<1);
   
   UInt uiPartStart          = uiPartIdx*uiQNumPart;
@@ -1822,7 +1822,7 @@
 *\param   uiCurrAbsIdxInLCU
 *\returns Char   reference QP value
 */
-Char TComDataCU::getRefQP( UInt uiCurrAbsIdxInLCU )
+UChar TComDataCU::getRefQP( UInt uiCurrAbsIdxInLCU )
 {
   UInt        lPartIdx, aPartIdx;
   TComDataCU* cULeft  = getQpMinCuLeft ( lPartIdx, m_uiAbsIdxInLCU + uiCurrAbsIdxInLCU );
@@ -1842,7 +1842,7 @@
   return iLastValidPartIdx;
 }
 
-Char TComDataCU::getLastCodedQP( UInt uiAbsPartIdx )
+UChar TComDataCU::getLastCodedQP( UInt uiAbsPartIdx )
 {
   UInt uiQUPartIdxMask = ~((1<<((g_uiMaxCUDepth - getSlice()->getPPS()->getMaxCuDQPDepth())<<1))-1);
   Int iLastValidPartIdx = getLastValidPartIdx( uiAbsPartIdx&uiQUPartIdxMask );
@@ -2330,12 +2330,12 @@
 
 Void TComDataCU::setMVPIdxSubParts( Int iMVPIdx, RefPicList eRefPicList, UInt uiAbsPartIdx, UInt uiPartIdx, UInt uiDepth )
 {
-  setSubPart<Char>( iMVPIdx, m_apiMVPIdx[eRefPicList], uiAbsPartIdx, uiDepth, uiPartIdx );
+  setSubPart<UChar>( iMVPIdx, m_apiMVPIdx[eRefPicList], uiAbsPartIdx, uiDepth, uiPartIdx );
 }
 
 Void TComDataCU::setMVPNumSubParts( Int iMVPNum, RefPicList eRefPicList, UInt uiAbsPartIdx, UInt uiPartIdx, UInt uiDepth )
 {
-  setSubPart<Char>( iMVPNum, m_apiMVPNum[eRefPicList], uiAbsPartIdx, uiDepth, uiPartIdx );
+  setSubPart<UChar>( iMVPNum, m_apiMVPNum[eRefPicList], uiAbsPartIdx, uiDepth, uiPartIdx );
 }
 
 
Index: source/Lib/TLibCommon/TComPicSym.h
===================================================================
--- source/Lib/TLibCommon/TComPicSym.h	(revision 2764)
+++ source/Lib/TLibCommon/TComPicSym.h	(working copy)
@@ -123,7 +123,7 @@
   UInt        getMinCUWidth()           { return m_uiMinCUWidth;                }
   UInt        getMinCUHeight()          { return m_uiMinCUHeight;               }
   UInt        getNumberOfCUsInFrame()   { return m_uiNumCUsInFrame;  }
-  TComDataCU*&  getCU( UInt uiCUAddr )  { return m_apcTComDataCU[uiCUAddr];     }
+  TComDataCU* getCU( UInt uiCUAddr )  { return m_apcTComDataCU[uiCUAddr];     }
   
   Void        setSlice(TComSlice* p, UInt i) { m_apcTComSlice[i] = p;           }
   UInt        getNumAllocatedSlice()    { return m_uiNumAllocatedSlice;         }
Index: source/Lib/TLibEncoder/TEncCavlc.h
===================================================================
--- source/Lib/TLibEncoder/TEncCavlc.h	(revision 2764)
+++ source/Lib/TLibEncoder/TEncCavlc.h	(working copy)
@@ -186,7 +186,6 @@
   
   Void xCodePredWeightTable          ( TComSlice* pcSlice );
   Void updateContextTables           ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true ) { return;   }
-  Void updateContextTables           ( SliceType eSliceType, Int iQp  )                          { return;   }
 
 #if !REMOVE_APS
   Void  codeAPSInitInfo(TComAPS* pcAPS);  //!< code APS flags before encoding SAO and ALF parameters
Index: source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.cpp
===================================================================
--- source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.cpp	(revision 2764)
+++ source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.cpp	(working copy)
@@ -2386,7 +2386,7 @@
  * \param yCbCr color component index
  * \param lambda 
  */
-inline Int64 TEncSampleAdaptiveOffset::estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo)
+Int64 TEncSampleAdaptiveOffset::estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo)
 {
   Int64 estDist = 0;
   Int classIdx;
@@ -2432,11 +2432,11 @@
   return estDist;
 }
 
-inline Int64 TEncSampleAdaptiveOffset::estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift)
+Int64 TEncSampleAdaptiveOffset::estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift)
 {
   return (( count*offset*offset-offsetOrg*offset*2 ) >> shift);
 }
-inline Int64 TEncSampleAdaptiveOffset::estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo )
+Int64 TEncSampleAdaptiveOffset::estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo )
 {
   //Clean up, best_q_offset.
   Int64 iterOffset, tempOffset;
Index: source/Lib/TLibEncoder/NALwrite.h
===================================================================
--- source/Lib/TLibEncoder/NALwrite.h	(revision 2764)
+++ source/Lib/TLibEncoder/NALwrite.h	(working copy)
@@ -83,7 +83,8 @@
 inline NALUnitEBSP::NALUnitEBSP(OutputNALUnit& nalu)
   : NALUnit(nalu)
 {
-  write(m_nalUnitData, nalu);
+  m_nalUnitData = new std::ostringstream();
+  write(*m_nalUnitData, nalu);
 }
 
 void copyNaluData(OutputNALUnit& naluDest, const OutputNALUnit& naluSrc);
Index: source/Lib/TLibEncoder/AnnexBwrite.h
===================================================================
--- source/Lib/TLibEncoder/AnnexBwrite.h	(revision 2764)
+++ source/Lib/TLibEncoder/AnnexBwrite.h	(working copy)
@@ -75,8 +75,9 @@
       out.write(start_code_prefix+1, 3);
       size += 3;
     }
-    out << nalu.m_nalUnitData.str();
-    size += unsigned(nalu.m_nalUnitData.str().size());
+    std::ostream *ud = nalu.m_nalUnitData;
+    out << ((std::ostringstream *)ud)->str();
+    size += unsigned(((std::ostringstream *)ud)->str().size());
 
     annexBsizes.push_back(size);
   }
Index: source/Lib/TLibEncoder/TEncGOP.cpp
===================================================================
--- source/Lib/TLibEncoder/TEncGOP.cpp	(revision 2764)
+++ source/Lib/TLibEncoder/TEncGOP.cpp	(working copy)
@@ -1682,7 +1682,9 @@
   unsigned numRBSPBytes = 0;
   for (AccessUnit::const_iterator it = accessUnit.begin(); it != accessUnit.end(); it++)
   {
-    unsigned numRBSPBytes_nal = unsigned((*it)->m_nalUnitData.str().size());
+    //unsigned numRBSPBytes_nal = unsigned((*it)->m_nalUnitData.str().size());
+    std::ostream *ud = (*it)->m_nalUnitData;
+    unsigned numRBSPBytes_nal = unsigned(((std::ostringstream*)ud)->str().size());
 #if VERBOSE_RATE
     printf("*** %6s numBytesInNALunit: %u\n", nalUnitTypeToString((*it)->m_nalUnitType), numRBSPBytes_nal);
 #endif
Index: source/Lib/TLibEncoder/TEncCavlc.cpp
===================================================================
--- source/Lib/TLibEncoder/TEncCavlc.cpp	(revision 2764)
+++ source/Lib/TLibEncoder/TEncCavlc.cpp	(working copy)
@@ -1428,7 +1428,7 @@
 
       for ( Int iRefIdx=0 ; iRefIdx<pcSlice->getNumRefIdx(eRefPicList) ; iRefIdx++ ) 
       {
-        pcSlice->getWpScaling(eRefPicList, iRefIdx, wp);
+        wp = pcSlice->getWpScaling(eRefPicList, iRefIdx);
         if ( !bDenomCoded ) 
         {
           Int iDeltaDenom;
@@ -1449,7 +1449,7 @@
       {
         for ( Int iRefIdx=0 ; iRefIdx<pcSlice->getNumRefIdx(eRefPicList) ; iRefIdx++ ) 
         {
-          pcSlice->getWpScaling(eRefPicList, iRefIdx, wp);
+          wp = pcSlice->getWpScaling(eRefPicList, iRefIdx);
           WRITE_FLAG( wp[1].bPresentFlag, "chroma_weight_lX_flag" );           // u(1): chroma_weight_lX_flag
           uiTotalSignalledWeightFlags += 2*wp[1].bPresentFlag;
         }
@@ -1457,7 +1457,7 @@
 
       for ( Int iRefIdx=0 ; iRefIdx<pcSlice->getNumRefIdx(eRefPicList) ; iRefIdx++ ) 
       {
-        pcSlice->getWpScaling(eRefPicList, iRefIdx, wp);
+        wp = pcSlice->getWpScaling(eRefPicList, iRefIdx);
 #endif
         if ( wp[0].bPresentFlag ) 
         {
Index: source/Lib/TLibEncoder/TEncSlice.cpp
===================================================================
--- source/Lib/TLibEncoder/TEncSlice.cpp	(revision 2764)
+++ source/Lib/TLibEncoder/TEncSlice.cpp	(working copy)
@@ -805,7 +805,7 @@
        uiCUAddr = rpcPic->getPicSym()->getCUOrderMap(++uiEncCUOrder) )
   {
     // initialize CU encoder
-    TComDataCU*& pcCU = rpcPic->getCU( uiCUAddr );
+    TComDataCU* pcCU = rpcPic->getCU( uiCUAddr );
     pcCU->initCU( rpcPic, uiCUAddr );
 
     if(m_pcCfg->getUseRateCtrl())
@@ -1235,7 +1235,7 @@
       }
     }
 
-    TComDataCU*& pcCU = rpcPic->getCU( uiCUAddr );    
+    TComDataCU* pcCU = rpcPic->getCU( uiCUAddr );    
 #if !SAO_LUM_CHROMA_ONOFF_FLAGS
     if ( pcSlice->getSPS()->getUseSAO() && pcSlice->getSaoEnabledFlag() )
 #else
Index: source/Lib/TLibEncoder/TEncBinCoderCABAC.cpp
===================================================================
--- source/Lib/TLibEncoder/TEncBinCoderCABAC.cpp	(revision 2764)
+++ source/Lib/TLibEncoder/TEncBinCoderCABAC.cpp	(working copy)
@@ -218,12 +218,12 @@
   m_uiBinsCoded += m_binCountIncrement;
   rcCtxModel.setBinsCoded( 1 );
   
-  UInt  uiLPS   = TComCABACTables::sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) & 3 ];
+  UInt  uiLPS   = sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) & 3 ];
   m_uiRange    -= uiLPS;
   
   if( binValue != rcCtxModel.getMps() )
   {
-    Int numBits = TComCABACTables::sm_aucRenormTable[ uiLPS >> 3 ];
+    Int numBits = sm_aucRenormTable[ uiLPS >> 3 ];
     m_uiLow     = ( m_uiLow + m_uiRange ) << numBits;
     m_uiRange   = uiLPS << numBits;
     rcCtxModel.updateLPS();
Index: source/Lib/TLibEncoder/TEncEntropy.h
===================================================================
--- source/Lib/TLibEncoder/TEncEntropy.h	(revision 2764)
+++ source/Lib/TLibEncoder/TEncEntropy.h	(working copy)
@@ -141,8 +141,7 @@
 #endif
   virtual Void estBit               (estBitsSbacStruct* pcEstBitsSbac, Int width, Int height, TextType eTType) = 0;
   
-  virtual Void updateContextTables ( SliceType eSliceType, Int iQp, Bool bExecuteFinish )   = 0;
-  virtual Void updateContextTables ( SliceType eSliceType, Int iQp )   = 0;
+  virtual Void updateContextTables ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true )   = 0;
 
 #if !REMOVE_APS
   virtual Void codeAPSInitInfo  (TComAPS* pcAPS)= 0;
@@ -226,8 +225,7 @@
 #endif
   Void encodeQtRootCbf         ( TComDataCU* pcCU, UInt uiAbsPartIdx );
   Void encodeQP                ( TComDataCU* pcCU, UInt uiAbsPartIdx, Bool bRD = false );
-  Void updateContextTables     ( SliceType eSliceType, Int iQp, Bool bExecuteFinish )   { m_pcEntropyCoderIf->updateContextTables( eSliceType, iQp, bExecuteFinish );     }
-  Void updateContextTables     ( SliceType eSliceType, Int iQp )                        { m_pcEntropyCoderIf->updateContextTables( eSliceType, iQp, true );               }
+  Void updateContextTables     ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true )   { m_pcEntropyCoderIf->updateContextTables( eSliceType, iQp, bExecuteFinish );     }
 
 #if !REMOVE_APS
   Void encodeAPSInitInfo          (TComAPS* pcAPS) {m_pcEntropyCoderIf->codeAPSInitInfo(pcAPS);}
Index: source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.h
===================================================================
--- source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.h	(revision 2764)
+++ source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.h	(working copy)
@@ -140,9 +140,9 @@
 #else
   Void rdoSaoUnit(Int rx, Int ry, SAOParam *saoParam, Int addr, Int addrUp, Int addrLeft, Int yCbCr, Double lambda);
 #endif
-  inline Int64 estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift);
-  inline Int64 estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo );
-  inline Int64 estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo);
+  Int64 estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift);
+  Int64 estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo );
+  Int64 estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo);
   Void setMaxNumOffsetsPerPic(Int iVal) {m_maxNumOffsetsPerPic = iVal; }
   Int  getMaxNumOffsetsPerPic() {return m_maxNumOffsetsPerPic; }
 };
Index: source/Lib/TLibEncoder/TEncSbac.h
===================================================================
--- source/Lib/TLibEncoder/TEncSbac.h	(revision 2764)
+++ source/Lib/TLibEncoder/TEncSbac.h	(working copy)
@@ -198,7 +198,6 @@
   Void estSignificantCoefficientsBit ( estBitsSbacStruct* pcEstBitsSbac, UInt uiCTXIdx, TextType eTType );
   
   Void updateContextTables           ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true  );
-  Void updateContextTables           ( SliceType eSliceType, Int iQp  ) { this->updateContextTables( eSliceType, iQp, true); };
   
   TEncBinIf* getEncBinIf()  { return m_pcBinIf; }
 private:
