diff -urN original/App/TAppDecoder/TAppDecTop.cpp source/App/TAppDecoder/TAppDecTop.cpp
--- original/App/TAppDecoder/TAppDecTop.cpp	2012-07-24 02:27:48.000000000 +0900
+++ source/App/TAppDecoder/TAppDecTop.cpp	2012-08-19 17:39:36.000000000 +0900
@@ -172,7 +172,9 @@
     }
     if (bNewPicture || !bitstreamFile)
     {
-      m_cTDecTop.executeDeblockAndAlf(uiPOC, pcListPic, m_iSkipFrame, m_iPOCLastDisplay);
+      TComList<TComPic*> *rpcListPic = m_cTDecTop.executeDeblockAndAlf(uiPOC, m_iSkipFrame, m_iPOCLastDisplay);
+      if (rpcListPic)
+        pcListPic = rpcListPic;
     }
 
     if( pcListPic )
diff -urN original/Lib/TLibCommon/NAL.h source/Lib/TLibCommon/NAL.h
--- original/Lib/TLibCommon/NAL.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/NAL.h	2012-08-18 14:40:09.000000000 +0900
@@ -88,7 +88,8 @@
  */
 struct NALUnitEBSP : public NALUnit
 {
-  std::ostringstream m_nalUnitData;
+//std::ostringstream m_nalUnitData;
+  std::ostream *m_nalUnitData;
 
   /**
    * convert the OutputNALUnit #nalu# into EBSP format by writing out
diff -urN original/Lib/TLibCommon/TComCABACTables.cpp source/Lib/TLibCommon/TComCABACTables.cpp
--- original/Lib/TLibCommon/TComCABACTables.cpp	2012-01-19 02:44:23.000000000 +0900
+++ source/Lib/TLibCommon/TComCABACTables.cpp	2012-08-17 16:41:56.000000000 +0900
@@ -40,7 +40,7 @@
 //! \ingroup TLibCommon
 //! \{
 
-const UChar TComCABACTables::sm_aucLPSTable[64][4] =
+const UChar sm_aucLPSTable[64][4] =
 {
   { 128, 176, 208, 240},
   { 128, 167, 197, 227},
@@ -108,7 +108,7 @@
   {   2,   2,   2,   2}
 };
 
-const UChar TComCABACTables::sm_aucRenormTable[32] =
+const UChar sm_aucRenormTable[32] =
 {
   6,  5,  4,  4,
   3,  3,  3,  3,
diff -urN original/Lib/TLibCommon/TComCABACTables.h source/Lib/TLibCommon/TComCABACTables.h
--- original/Lib/TLibCommon/TComCABACTables.h	2012-01-19 02:44:23.000000000 +0900
+++ source/Lib/TLibCommon/TComCABACTables.h	2012-08-17 16:41:37.000000000 +0900
@@ -47,13 +47,8 @@
  * \brief static class for CABAC tables
  */
 
-class TComCABACTables
-{
-public:
-  const static UChar  sm_aucLPSTable[64][4];
-  const static UChar  sm_aucRenormTable[32];
-};
-
+extern const UChar sm_aucLPSTable[64][4];
+extern const UChar sm_aucRenormTable[32];
 
 //! \}
 
diff -urN original/Lib/TLibCommon/TComDataCU.cpp source/Lib/TLibCommon/TComDataCU.cpp
--- original/Lib/TLibCommon/TComDataCU.cpp	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/TComDataCU.cpp	2012-08-22 21:35:58.000000000 +0900
@@ -1057,7 +1057,7 @@
 // It is used to predict for next part
 Void TComDataCU::copyToPic( UChar uhDepth )
 {
-  TComDataCU*& rpcCU = m_pcPic->getCU( m_uiCUAddr );
+  TComDataCU* rpcCU = m_pcPic->getCU( m_uiCUAddr );
   
   rpcCU->getTotalCost()       = m_dTotalCost;
   rpcCU->getTotalDistortion() = m_uiTotalDistortion;
@@ -1135,7 +1135,7 @@
 
 Void TComDataCU::copyToPic( UChar uhDepth, UInt uiPartIdx, UInt uiPartDepth )
 {
-  TComDataCU*&  rpcCU       = m_pcPic->getCU( m_uiCUAddr );
+  TComDataCU*   rpcCU       = m_pcPic->getCU( m_uiCUAddr );
   UInt          uiQNumPart  = m_uiNumPartition>>(uiPartDepth<<1);
   
   UInt uiPartStart          = uiPartIdx*uiQNumPart;
diff -urN original/Lib/TLibCommon/TComDataCU.h source/Lib/TLibCommon/TComDataCU.h
--- original/Lib/TLibCommon/TComDataCU.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/TComDataCU.h	2012-08-25 14:31:36.000000000 +0900
@@ -263,8 +263,8 @@
   
   TComPic*      getPic                ()                        { return m_pcPic;           }
   TComSlice*    getSlice              ()                        { return m_pcSlice;         }
-  UInt&         getAddr               ()                        { return m_uiCUAddr;        }
-  UInt&         getZorderIdxInCU      ()                        { return m_uiAbsIdxInLCU; }
+  UInt          getAddr               ()                        { return m_uiCUAddr;        }
+  UInt          getZorderIdxInCU      ()                        { return m_uiAbsIdxInLCU; }
   UInt          getSCUAddr            ();
   UInt          getCUPelX             ()                        { return m_uiCUPelX;        }
   UInt          getCUPelY             ()                        { return m_uiCUPelY;        }
@@ -341,18 +341,18 @@
   
   TComCUMvField* getCUMvField         ( RefPicList e )          { return  &m_acCUMvField[e];  }
   
-  TCoeff*&      getCoeffY             ()                        { return m_pcTrCoeffY;        }
-  TCoeff*&      getCoeffCb            ()                        { return m_pcTrCoeffCb;       }
-  TCoeff*&      getCoeffCr            ()                        { return m_pcTrCoeffCr;       }
+  TCoeff*       getCoeffY             ()                        { return m_pcTrCoeffY;        }
+  TCoeff*       getCoeffCb            ()                        { return m_pcTrCoeffCb;       }
+  TCoeff*       getCoeffCr            ()                        { return m_pcTrCoeffCr;       }
 #if ADAPTIVE_QP_SELECTION
-  Int*&         getArlCoeffY          ()                        { return m_pcArlCoeffY;       }
-  Int*&         getArlCoeffCb         ()                        { return m_pcArlCoeffCb;      }
-  Int*&         getArlCoeffCr         ()                        { return m_pcArlCoeffCr;      }
+  Int*          getArlCoeffY          ()                        { return m_pcArlCoeffY;       }
+  Int*          getArlCoeffCb         ()                        { return m_pcArlCoeffCb;      }
+  Int*          getArlCoeffCr         ()                        { return m_pcArlCoeffCr;      }
 #endif
   
-  Pel*&         getPCMSampleY         ()                        { return m_pcIPCMSampleY;     }
-  Pel*&         getPCMSampleCb        ()                        { return m_pcIPCMSampleCb;    }
-  Pel*&         getPCMSampleCr        ()                        { return m_pcIPCMSampleCr;    }
+  Pel*          getPCMSampleY         ()                        { return m_pcIPCMSampleY;     }
+  Pel*          getPCMSampleCb        ()                        { return m_pcIPCMSampleCb;    }
+  Pel*          getPCMSampleCr        ()                        { return m_pcIPCMSampleCr;    }
 
   UChar         getCbf    ( UInt uiIdx, TextType eType )                  { return m_puhCbf[g_aucConvertTxtTypeToIdx[eType]][uiIdx];  }
   UChar*        getCbf    ( TextType eType )                              { return m_puhCbf[g_aucConvertTxtTypeToIdx[eType]];         }
@@ -548,7 +548,7 @@
   Double&       getTotalCost()                  { return m_dTotalCost;        }
   UInt&         getTotalDistortion()            { return m_uiTotalDistortion; }
   UInt&         getTotalBits()                  { return m_uiTotalBits;       }
-  UInt&         getTotalNumPart()               { return m_uiNumPartition;    }
+  UInt          getTotalNumPart()               { return m_uiNumPartition;    }
 
   UInt          getCoefScanIdx(UInt uiAbsPartIdx, UInt uiWidth, Bool bIsLuma, Bool bIsIntra);
 
diff -urN original/Lib/TLibCommon/TComList.h source/Lib/TLibCommon/TComList.h
--- original/Lib/TLibCommon/TComList.h	2012-01-19 02:44:23.000000000 +0900
+++ source/Lib/TLibCommon/TComList.h	2012-08-19 06:14:49.000000000 +0900
@@ -67,7 +67,7 @@
   {
     if( ! rcTComList.empty() )
     {
-      insert( this->end(), rcTComList.begin(), rcTComList.end());
+      this->insert( this->end(), rcTComList.begin(), rcTComList.end());
     }
     return *this;
   } // leszek
@@ -104,10 +104,10 @@
     }
   }
   
-  TComIterator find( const C& rcT ) // leszek
+/*TComIterator find( const C& rcT ) // leszek
   {
-    return find( this->begin(), this->end(), rcT );
-  }
+    return this->find( this->begin(), this->end(), rcT );
+  }*/
 };
 
 //! \}
diff -urN original/Lib/TLibCommon/TComPic.h source/Lib/TLibCommon/TComPic.h
--- original/Lib/TLibCommon/TComPic.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/TComPic.h	2012-08-22 21:33:26.000000000 +0900
@@ -106,7 +106,7 @@
   TComPicSym*   getPicSym()           { return  m_apcPicSym;    }
   TComSlice*    getSlice(Int i)       { return  m_apcPicSym->getSlice(i);  }
   Int           getPOC()              { return  m_apcPicSym->getSlice(m_uiCurrSliceIdx)->getPOC();  }
-  TComDataCU*&  getCU( UInt uiCUAddr )  { return  m_apcPicSym->getCU( uiCUAddr ); }
+  TComDataCU*   getCU( UInt uiCUAddr )  { return  m_apcPicSym->getCU( uiCUAddr ); }
   
   TComPicYuv*   getPicYuvOrg()        { return  m_apcPicYuv[0]; }
   TComPicYuv*   getPicYuvRec()        { return  m_apcPicYuv[1]; }
diff -urN original/Lib/TLibCommon/TComPicSym.h source/Lib/TLibCommon/TComPicSym.h
--- original/Lib/TLibCommon/TComPicSym.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/TComPicSym.h	2012-08-22 21:33:43.000000000 +0900
@@ -123,7 +123,7 @@
   UInt        getMinCUWidth()           { return m_uiMinCUWidth;                }
   UInt        getMinCUHeight()          { return m_uiMinCUHeight;               }
   UInt        getNumberOfCUsInFrame()   { return m_uiNumCUsInFrame;  }
-  TComDataCU*&  getCU( UInt uiCUAddr )  { return m_apcTComDataCU[uiCUAddr];     }
+  TComDataCU* getCU( UInt uiCUAddr )  { return m_apcTComDataCU[uiCUAddr];     }
   
   Void        setSlice(TComSlice* p, UInt i) { m_apcTComSlice[i] = p;           }
   UInt        getNumAllocatedSlice()    { return m_uiNumAllocatedSlice;         }
diff -urN original/Lib/TLibCommon/TComRom.cpp source/Lib/TLibCommon/TComRom.cpp
--- original/Lib/TLibCommon/TComRom.cpp	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/TComRom.cpp	2012-08-25 11:18:03.000000000 +0900
@@ -179,7 +179,7 @@
 
 UInt g_auiPUOffset[8] = { 0, 8, 4, 4, 2, 10, 1, 5};
 
-Void initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt*& rpuiCurrIdx )
+UInt *initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt* rpuiCurrIdx )
 {
   Int iStride = 1 << ( iMaxDepth - 1 );
   
@@ -191,11 +191,13 @@
   else
   {
     Int iStep = iStride >> iDepth;
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal,                     rpuiCurrIdx );
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep,               rpuiCurrIdx );
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride,       rpuiCurrIdx );
-    initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride+iStep, rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal,                     rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep,               rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride,       rpuiCurrIdx );
+    rpuiCurrIdx = initZscanToRaster( iMaxDepth, iDepth+1, uiStartVal+iStep*iStride+iStep, rpuiCurrIdx );
   }
+
+  return rpuiCurrIdx;
 }
 
 Void initRasterToZscan ( UInt uiMaxCUWidth, UInt uiMaxCUHeight, UInt uiMaxDepth )
@@ -402,6 +404,8 @@
   {84,  -29,   -74,   55},
   {55,  -84,    74,  -29},
 };
+const UChar g_aucDCTDSTMode_Vert[NUM_INTRA_MODE] = {};
+const UChar g_aucDCTDSTMode_Hor[NUM_INTRA_MODE] = {};
 
 
 // ====================================================================================================================
@@ -432,6 +436,10 @@
 };
 #endif // FAST_UDI_USE_MPM
 
+const UChar g_aucIntraModeNumAng[7] = {};
+const UChar g_aucIntraModeBitsAng[7] = {};
+const UChar g_aucAngIntraModeOrder[NUM_INTRA_MODE] = {};
+
 // chroma
 
 const UChar g_aucConvertTxtTypeToIdx[4] = { 0, 1, 1, 2 };
@@ -826,6 +834,10 @@
   20,24,25,28,33,41,54,71,
   24,25,28,33,41,54,71,91
 };
+Int g_quantIntraDefault16x16[256] = {};
+Int g_quantIntraDefault32x32[1024] = {};
+Int g_quantInterDefault16x16[256] = {};
+Int g_quantInterDefault32x32[1024] = {};
 UInt g_scalingListSize   [4] = {16,64,256,1024}; 
 UInt g_scalingListSizeX  [4] = { 4, 8, 16,  32};
 UInt g_scalingListNum[SCALING_LIST_SIZE_NUM]={6,6,6,2};
diff -urN original/Lib/TLibCommon/TComRom.h source/Lib/TLibCommon/TComRom.h
--- original/Lib/TLibCommon/TComRom.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/TComRom.h	2012-08-25 11:17:57.000000000 +0900
@@ -53,7 +53,8 @@
 #define     MAX_CU_DEPTH            7                           // log2(LCUSize)
 #define     MAX_CU_SIZE             (1<<(MAX_CU_DEPTH))         // maximum allowable size of CU
 #define     MIN_PU_SIZE             4
-#define     MAX_NUM_SPU_W           (MAX_CU_SIZE/MIN_PU_SIZE)   // maximum number of SPU in horizontal line
+//#define     MAX_NUM_SPU_W           (MAX_CU_SIZE/MIN_PU_SIZE)   // maximum number of SPU in horizontal line
+#define     MAX_NUM_SPU_W           (32)   // maximum number of SPU in horizontal line
 
 // ====================================================================================================================
 // Initialize / destroy functions
@@ -75,7 +76,7 @@
 extern       UInt   g_auiRasterToZscan[ MAX_NUM_SPU_W*MAX_NUM_SPU_W ];
 extern       UInt   g_motionRefer[ MAX_NUM_SPU_W*MAX_NUM_SPU_W ];
 
-Void         initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt*& rpuiCurrIdx );
+UInt        *initZscanToRaster ( Int iMaxDepth, Int iDepth, UInt uiStartVal, UInt* rpuiCurrIdx );
 Void         initRasterToZscan ( UInt uiMaxCUWidth, UInt uiMaxCUHeight, UInt uiMaxDepth         );
 
 Void          initMotionReferIdx ( UInt uiMaxCUWidth, UInt uiMaxCUHeight, UInt uiMaxDepth );
diff -urN original/Lib/TLibCommon/TComSlice.h source/Lib/TLibCommon/TComSlice.h
--- original/Lib/TLibCommon/TComSlice.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibCommon/TComSlice.h	2012-08-18 00:05:40.000000000 +0900
@@ -496,7 +496,7 @@
   Void setScalingListFlag       ( Bool b ) { m_scalingListEnabledFlag  = b;       }
   Bool getScalingListPresentFlag()         { return m_scalingListPresentFlag;     }
   Void setScalingListPresentFlag( Bool b ) { m_scalingListPresentFlag  = b;       }
-  Void setScalingList      ( TComScalingList *scalingList);
+  Void setScalingList      ( TComScalingList *scalingList) {}
   TComScalingList* getScalingList ()       { return m_scalingList; }               //!< get ScalingList class pointer in SPS
   UInt getMaxDecPicBuffering  (UInt tlayer)            { return m_uiMaxDecPicBuffering[tlayer]; }
   Void setMaxDecPicBuffering  ( UInt ui, UInt tlayer ) { m_uiMaxDecPicBuffering[tlayer] = ui;   }
@@ -517,8 +517,8 @@
   TComRefPicListModification();
   virtual ~TComRefPicListModification();
   
-  Void  create                    ();
-  Void  destroy                   ();
+  Void  create                    () {}
+  Void  destroy                   () {}
 
   Bool       getRefPicListModificationFlagL0() { return m_bRefPicListModificationFlagL0; }
   Void       setRefPicListModificationFlagL0(Bool flag) { m_bRefPicListModificationFlagL0 = flag; }
@@ -718,7 +718,7 @@
   Int      getLoopFilterTcOffset()             {return m_loopFilterTcOffsetDiv2; }     //!< get tc offset for deblocking filter
   Bool     getScalingListPresentFlag()         { return m_scalingListPresentFlag;     }
   Void     setScalingListPresentFlag( Bool b ) { m_scalingListPresentFlag  = b;       }
-  Void     setScalingList      ( TComScalingList *scalingList);
+  Void     setScalingList      ( TComScalingList *scalingList) {}
   TComScalingList* getScalingList ()          { return m_scalingList; }         //!< get ScalingList class pointer in PPS
   UInt getLog2ParallelMergeLevelMinus2      ()                    { return m_log2ParallelMergeLevelMinus2; }
   Void setLog2ParallelMergeLevelMinus2      (UInt mrgLevel)       { m_log2ParallelMergeLevelMinus2 = mrgLevel; }
@@ -1106,7 +1106,7 @@
   Void setTLayer             ( UInt uiTLayer )             { m_uiTLayer = uiTLayer;                  }
 
   Void setTLayerInfo( UInt uiTLayer );
-  Void decodingMarking( TComList<TComPic*>& rcListPic, Int iGOPSIze, Int& iMaxRefPicNum ); 
+  Void decodingMarking( TComList<TComPic*>& rcListPic, Int iGOPSIze, Int& iMaxRefPicNum ) {}
   Void applyReferencePictureSet( TComList<TComPic*>& rcListPic, TComReferencePictureSet *RPSList);
   Bool isTemporalLayerSwitchingPoint( TComList<TComPic*>& rcListPic, TComReferencePictureSet *RPSList);
   Int       checkThatAllRefPicsAreAvailable( TComList<TComPic*>& rcListPic, TComReferencePictureSet *pReferencePictureSet, Bool printErrors, Int pocRandomAccess = 0);
diff -urN original/Lib/TLibDecoder/TDecBinCoderCABAC.cpp source/Lib/TLibDecoder/TDecBinCoderCABAC.cpp
--- original/Lib/TLibDecoder/TDecBinCoderCABAC.cpp	2012-06-26 16:19:43.000000000 +0900
+++ source/Lib/TLibDecoder/TDecBinCoderCABAC.cpp	2012-08-17 16:42:23.000000000 +0900
@@ -105,7 +105,7 @@
 Void
 TDecBinCABAC::decodeBin( UInt& ruiBin, ContextModel &rcCtxModel )
 {
-  UInt uiLPS = TComCABACTables::sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) - 4 ];
+  UInt uiLPS = sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) - 4 ];
   m_uiRange -= uiLPS;
   UInt scaledRange = m_uiRange << 7;
   
@@ -132,7 +132,7 @@
   else
   {
     // LPS path
-    Int numBits = TComCABACTables::sm_aucRenormTable[ uiLPS >> 3 ];
+    Int numBits = sm_aucRenormTable[ uiLPS >> 3 ];
     m_uiValue   = ( m_uiValue - scaledRange ) << numBits;
     m_uiRange   = uiLPS << numBits;
     ruiBin      = 1 - rcCtxModel.getMps();
diff -urN original/Lib/TLibDecoder/TDecCAVLC.cpp source/Lib/TLibDecoder/TDecCAVLC.cpp
--- original/Lib/TLibDecoder/TDecCAVLC.cpp	2012-08-07 03:11:44.000000000 +0900
+++ source/Lib/TLibDecoder/TDecCAVLC.cpp	2012-08-22 19:51:14.000000000 +0900
@@ -802,7 +802,7 @@
   return;
 }
 
-Void TDecCavlc::parseSliceHeader (TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager)
+Void TDecCavlc::parseSliceHeader (TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager)
 {
   UInt  uiCode;
   Int   iCode;
diff -urN original/Lib/TLibDecoder/TDecCAVLC.h source/Lib/TLibDecoder/TDecCAVLC.h
--- original/Lib/TLibDecoder/TDecCAVLC.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibDecoder/TDecCAVLC.h	2012-08-22 19:51:21.000000000 +0900
@@ -108,7 +108,7 @@
 #if !REMOVE_APS
   Void  parseAPS            ( TComAPS* pAPS );
 #endif
-  Void  parseSliceHeader    ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager);
+  Void  parseSliceHeader    ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager);
   Void  parseTerminatingBit ( UInt& ruiBit );
   
   Void  parseMVPIdx         ( Int& riMVPIdx );
diff -urN original/Lib/TLibDecoder/TDecEntropy.h source/Lib/TLibDecoder/TDecEntropy.h
--- original/Lib/TLibDecoder/TDecEntropy.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibDecoder/TDecEntropy.h	2012-08-22 19:51:48.000000000 +0900
@@ -75,7 +75,7 @@
 #endif
   virtual void parseSEI(SEImessages&) = 0;
 
-  virtual Void parseSliceHeader          ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager)       = 0;
+  virtual Void parseSliceHeader          ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager)       = 0;
 
   virtual Void  parseTerminatingBit       ( UInt& ruilsLast )                                     = 0;
   
@@ -149,7 +149,7 @@
 #endif
   void decodeSEI(SEImessages& seis) { m_pcEntropyDecoderIf->parseSEI(seis); }
 
-  Void    decodeSliceHeader           ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager)  { m_pcEntropyDecoderIf->parseSliceHeader(rpcSlice, parameterSetManager);         }
+  Void    decodeSliceHeader           ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager)  { m_pcEntropyDecoderIf->parseSliceHeader(rpcSlice, parameterSetManager);         }
 
   Void    decodeTerminatingBit        ( UInt& ruiIsLast )       { m_pcEntropyDecoderIf->parseTerminatingBit(ruiIsLast);     }
   
diff -urN original/Lib/TLibDecoder/TDecGop.cpp source/Lib/TLibDecoder/TDecGop.cpp
--- original/Lib/TLibDecoder/TDecGop.cpp	2012-08-07 03:11:44.000000000 +0900
+++ source/Lib/TLibDecoder/TDecGop.cpp	2012-08-22 23:09:50.000000000 +0900
@@ -109,7 +109,7 @@
 // Public member functions
 // ====================================================================================================================
 
-Void TDecGop::decompressSlice(TComInputBitstream* pcBitstream, TComPic*& rpcPic)
+Void TDecGop::decompressSlice(TComInputBitstream* pcBitstream, TComPic* rpcPic)
 {
   TComSlice*  pcSlice = rpcPic->getSlice(rpcPic->getCurrSliceIdx());
   // Table of extracted substreams.
@@ -198,7 +198,7 @@
   m_dDecTime += (double)(clock()-iBeforeTime) / CLOCKS_PER_SEC;
 }
 
-Void TDecGop::filterPicture(TComPic*& rpcPic)
+Void TDecGop::filterPicture(TComPic* rpcPic)
 {
   TComSlice*  pcSlice = rpcPic->getSlice(rpcPic->getCurrSliceIdx());
 
diff -urN original/Lib/TLibDecoder/TDecGop.h source/Lib/TLibDecoder/TDecGop.h
--- original/Lib/TLibDecoder/TDecGop.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibDecoder/TDecGop.h	2012-08-22 23:09:56.000000000 +0900
@@ -113,8 +113,8 @@
   Void  create  ();
   Void  destroy ();
 //  Void  decompressGop(TComInputBitstream* pcBitstream, TComPic*& rpcPic, Bool bExecuteDeblockAndAlf );
-  Void  decompressSlice(TComInputBitstream* pcBitstream, TComPic*& rpcPic );
-  Void  filterPicture  (TComPic*& rpcPic );
+  Void  decompressSlice(TComInputBitstream* pcBitstream, TComPic* rpcPic );
+  Void  filterPicture  (TComPic* rpcPic );
   Void  setGopSize( Int i) { m_iGopSize = i; }
 
   void setPictureDigestEnabled(Int enabled) { m_pictureDigestEnabled = enabled; }
diff -urN original/Lib/TLibDecoder/TDecSbac.h source/Lib/TLibDecoder/TDecSbac.h
--- original/Lib/TLibDecoder/TDecSbac.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibDecoder/TDecSbac.h	2012-08-22 19:52:06.000000000 +0900
@@ -85,7 +85,7 @@
 #endif
   void parseSEI(SEImessages&) {}
 
-  Void  parseSliceHeader          ( TComSlice*& rpcSlice, ParameterSetManagerDecoder *parameterSetManager) {}
+  Void  parseSliceHeader          ( TComSlice* rpcSlice, ParameterSetManagerDecoder *parameterSetManager) {}
   Void  parseTerminatingBit       ( UInt& ruiBit );
   Void  parseMVPIdx               ( Int& riMVPIdx          );
   Void  parseSaoMaxUvlc           ( UInt& val, UInt maxSymbol );
diff -urN original/Lib/TLibDecoder/TDecSlice.cpp source/Lib/TLibDecoder/TDecSlice.cpp
--- original/Lib/TLibDecoder/TDecSlice.cpp	2012-08-07 03:11:44.000000000 +0900
+++ source/Lib/TLibDecoder/TDecSlice.cpp	2012-08-22 21:11:19.000000000 +0900
@@ -90,7 +90,7 @@
   m_pcCuDecoder       = pcCuDecoder;
 }
 
-Void TDecSlice::decompressSlice(TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams, TComPic*& rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders)
+Void TDecSlice::decompressSlice(TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams, TComPic* rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders)
 {
   TComDataCU* pcCU;
   UInt        uiIsLast = 0;
diff -urN original/Lib/TLibDecoder/TDecSlice.h source/Lib/TLibDecoder/TDecSlice.h
--- original/Lib/TLibDecoder/TDecSlice.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibDecoder/TDecSlice.h	2012-08-22 21:11:29.000000000 +0900
@@ -79,7 +79,7 @@
   Void  create            ( TComSlice* pcSlice, Int iWidth, Int iHeight, UInt uiMaxWidth, UInt uiMaxHeight, UInt uiMaxDepth );
   Void  destroy           ();
   
-  Void  decompressSlice   ( TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams,   TComPic*& rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders );
+  Void  decompressSlice   ( TComInputBitstream* pcBitstream, TComInputBitstream** ppcSubstreams,   TComPic* rpcPic, TDecSbac* pcSbacDecoder, TDecSbac* pcSbacDecoders );
 };
 
 
diff -urN original/Lib/TLibDecoder/TDecTop.cpp source/Lib/TLibDecoder/TDecTop.cpp
--- original/Lib/TLibDecoder/TDecTop.cpp	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibDecoder/TDecTop.cpp	2012-08-19 17:55:13.000000000 +0900
@@ -191,14 +191,15 @@
 #endif
 }
 
-Void TDecTop::executeDeblockAndAlf(UInt& ruiPOC, TComList<TComPic*>*& rpcListPic, Int& iSkipFrame, Int& iPOCLastDisplay)
+TComList<TComPic*> *TDecTop::executeDeblockAndAlf(UInt& ruiPOC, Int& iSkipFrame, Int& iPOCLastDisplay)
 {
   if (!m_pcPic)
   {
     /* nothing to deblock */
-    return;
+    return NULL;
   }
-  
+
+  TComList<TComPic*>* rpcListPic = NULL;
   TComPic*&   pcPic         = m_pcPic;
 
   // Execute Deblock and ALF only + Cleanup
@@ -211,7 +212,7 @@
   m_cCuDecoder.destroy();        
   m_bFirstSliceInPicture  = true;
 
-  return;
+  return rpcListPic;
 }
 
 Void TDecTop::xCreateLostPicture(Int iLostPoc) 
diff -urN original/Lib/TLibDecoder/TDecTop.h source/Lib/TLibDecoder/TDecTop.h
--- original/Lib/TLibDecoder/TDecTop.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibDecoder/TDecTop.h	2012-08-19 17:34:33.000000000 +0900
@@ -116,7 +116,7 @@
   
   Void  deletePicBuffer();
 
-  Void executeDeblockAndAlf(UInt& ruiPOC, TComList<TComPic*>*& rpcListPic, Int& iSkipFrame,  Int& iPOCLastDisplay);
+  TComList<TComPic*> *executeDeblockAndAlf(UInt& ruiPOC, Int& iSkipFrame,  Int& iPOCLastDisplay);
 
 protected:
   Void  xGetNewPicBuffer  (TComSlice* pcSlice, TComPic*& rpcPic);
diff -urN original/Lib/TLibEncoder/AnnexBwrite.h source/Lib/TLibEncoder/AnnexBwrite.h
--- original/Lib/TLibEncoder/AnnexBwrite.h	2012-03-31 07:40:48.000000000 +0900
+++ source/Lib/TLibEncoder/AnnexBwrite.h	2012-08-17 23:27:49.000000000 +0900
@@ -75,8 +75,9 @@
       out.write(start_code_prefix+1, 3);
       size += 3;
     }
-    out << nalu.m_nalUnitData.str();
-    size += unsigned(nalu.m_nalUnitData.str().size());
+    std::ostream *ud = nalu.m_nalUnitData;
+    out << ((std::ostringstream *)ud)->str();
+    size += unsigned(((std::ostringstream *)ud)->str().size());
 
     annexBsizes.push_back(size);
   }
diff -urN original/Lib/TLibEncoder/NALwrite.h source/Lib/TLibEncoder/NALwrite.h
--- original/Lib/TLibEncoder/NALwrite.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibEncoder/NALwrite.h	2012-08-18 14:42:52.000000000 +0900
@@ -83,7 +83,8 @@
 inline NALUnitEBSP::NALUnitEBSP(OutputNALUnit& nalu)
   : NALUnit(nalu)
 {
-  write(m_nalUnitData, nalu);
+  m_nalUnitData = new std::ostringstream();
+  write(*m_nalUnitData, nalu);
 }
 
 void copyNaluData(OutputNALUnit& naluDest, const OutputNALUnit& naluSrc);
diff -urN original/Lib/TLibEncoder/TEncBinCoderCABAC.cpp source/Lib/TLibEncoder/TEncBinCoderCABAC.cpp
--- original/Lib/TLibEncoder/TEncBinCoderCABAC.cpp	2012-06-26 16:19:43.000000000 +0900
+++ source/Lib/TLibEncoder/TEncBinCoderCABAC.cpp	2012-08-17 16:42:40.000000000 +0900
@@ -218,12 +218,12 @@
   m_uiBinsCoded += m_binCountIncrement;
   rcCtxModel.setBinsCoded( 1 );
   
-  UInt  uiLPS   = TComCABACTables::sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) & 3 ];
+  UInt  uiLPS   = sm_aucLPSTable[ rcCtxModel.getState() ][ ( m_uiRange >> 6 ) & 3 ];
   m_uiRange    -= uiLPS;
   
   if( binValue != rcCtxModel.getMps() )
   {
-    Int numBits = TComCABACTables::sm_aucRenormTable[ uiLPS >> 3 ];
+    Int numBits = sm_aucRenormTable[ uiLPS >> 3 ];
     m_uiLow     = ( m_uiLow + m_uiRange ) << numBits;
     m_uiRange   = uiLPS << numBits;
     rcCtxModel.updateLPS();
diff -urN original/Lib/TLibEncoder/TEncCavlc.h source/Lib/TLibEncoder/TEncCavlc.h
--- original/Lib/TLibEncoder/TEncCavlc.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibEncoder/TEncCavlc.h	2012-08-17 16:27:23.000000000 +0900
@@ -186,7 +186,6 @@
   
   Void xCodePredWeightTable          ( TComSlice* pcSlice );
   Void updateContextTables           ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true ) { return;   }
-  Void updateContextTables           ( SliceType eSliceType, Int iQp  )                          { return;   }
 
 #if !REMOVE_APS
   Void  codeAPSInitInfo(TComAPS* pcAPS);  //!< code APS flags before encoding SAO and ALF parameters
diff -urN original/Lib/TLibEncoder/TEncEntropy.h source/Lib/TLibEncoder/TEncEntropy.h
--- original/Lib/TLibEncoder/TEncEntropy.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibEncoder/TEncEntropy.h	2012-08-17 16:27:04.000000000 +0900
@@ -141,8 +141,7 @@
 #endif
   virtual Void estBit               (estBitsSbacStruct* pcEstBitsSbac, Int width, Int height, TextType eTType) = 0;
   
-  virtual Void updateContextTables ( SliceType eSliceType, Int iQp, Bool bExecuteFinish )   = 0;
-  virtual Void updateContextTables ( SliceType eSliceType, Int iQp )   = 0;
+  virtual Void updateContextTables ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true )   = 0;
 
 #if !REMOVE_APS
   virtual Void codeAPSInitInfo  (TComAPS* pcAPS)= 0;
@@ -226,8 +225,7 @@
 #endif
   Void encodeQtRootCbf         ( TComDataCU* pcCU, UInt uiAbsPartIdx );
   Void encodeQP                ( TComDataCU* pcCU, UInt uiAbsPartIdx, Bool bRD = false );
-  Void updateContextTables     ( SliceType eSliceType, Int iQp, Bool bExecuteFinish )   { m_pcEntropyCoderIf->updateContextTables( eSliceType, iQp, bExecuteFinish );     }
-  Void updateContextTables     ( SliceType eSliceType, Int iQp )                        { m_pcEntropyCoderIf->updateContextTables( eSliceType, iQp, true );               }
+  Void updateContextTables     ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true )   { m_pcEntropyCoderIf->updateContextTables( eSliceType, iQp, bExecuteFinish );     }
 
 #if !REMOVE_APS
   Void encodeAPSInitInfo          (TComAPS* pcAPS) {m_pcEntropyCoderIf->codeAPSInitInfo(pcAPS);}
diff -urN original/Lib/TLibEncoder/TEncGOP.cpp source/Lib/TLibEncoder/TEncGOP.cpp
--- original/Lib/TLibEncoder/TEncGOP.cpp	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibEncoder/TEncGOP.cpp	2012-08-17 23:25:46.000000000 +0900
@@ -1682,7 +1682,9 @@
   unsigned numRBSPBytes = 0;
   for (AccessUnit::const_iterator it = accessUnit.begin(); it != accessUnit.end(); it++)
   {
-    unsigned numRBSPBytes_nal = unsigned((*it)->m_nalUnitData.str().size());
+    //unsigned numRBSPBytes_nal = unsigned((*it)->m_nalUnitData.str().size());
+    std::ostream *ud = (*it)->m_nalUnitData;
+    unsigned numRBSPBytes_nal = unsigned(((std::ostringstream*)ud)->str().size());
 #if VERBOSE_RATE
     printf("*** %6s numBytesInNALunit: %u\n", nalUnitTypeToString((*it)->m_nalUnitType), numRBSPBytes_nal);
 #endif
diff -urN original/Lib/TLibEncoder/TEncSampleAdaptiveOffset.cpp source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.cpp
--- original/Lib/TLibEncoder/TEncSampleAdaptiveOffset.cpp	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.cpp	2012-08-17 23:46:37.000000000 +0900
@@ -2386,7 +2386,7 @@
  * \param yCbCr color component index
  * \param lambda 
  */
-inline Int64 TEncSampleAdaptiveOffset::estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo)
+Int64 TEncSampleAdaptiveOffset::estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo)
 {
   Int64 estDist = 0;
   Int classIdx;
@@ -2432,11 +2432,11 @@
   return estDist;
 }
 
-inline Int64 TEncSampleAdaptiveOffset::estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift)
+Int64 TEncSampleAdaptiveOffset::estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift)
 {
   return (( count*offset*offset-offsetOrg*offset*2 ) >> shift);
 }
-inline Int64 TEncSampleAdaptiveOffset::estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo )
+Int64 TEncSampleAdaptiveOffset::estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo )
 {
   //Clean up, best_q_offset.
   Int64 iterOffset, tempOffset;
diff -urN original/Lib/TLibEncoder/TEncSampleAdaptiveOffset.h source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.h
--- original/Lib/TLibEncoder/TEncSampleAdaptiveOffset.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibEncoder/TEncSampleAdaptiveOffset.h	2012-08-17 23:46:19.000000000 +0900
@@ -140,9 +140,9 @@
 #else
   Void rdoSaoUnit(Int rx, Int ry, SAOParam *saoParam, Int addr, Int addrUp, Int addrLeft, Int yCbCr, Double lambda);
 #endif
-  inline Int64 estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift);
-  inline Int64 estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo );
-  inline Int64 estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo);
+  Int64 estSaoDist(Int64 count, Int64 offset, Int64 offsetOrg, Int shift);
+  Int64 estIterOffset(Int typeIdx, Int classIdx, double lambda, Int64 offsetInput, Int64 count, Int64 offsetOrg, Int shift, Int bitIncrease, Int *currentDistortionTableBo, Double *currentRdCostTableBo );
+  Int64 estSaoTypeDist(Int compIdx, Int typeIdx, Int shift, Double lambda, Int *currentDistortionTableBo, Double *currentRdCostTableBo);
   Void setMaxNumOffsetsPerPic(Int iVal) {m_maxNumOffsetsPerPic = iVal; }
   Int  getMaxNumOffsetsPerPic() {return m_maxNumOffsetsPerPic; }
 };
diff -urN original/Lib/TLibEncoder/TEncSbac.h source/Lib/TLibEncoder/TEncSbac.h
--- original/Lib/TLibEncoder/TEncSbac.h	2012-08-05 06:08:01.000000000 +0900
+++ source/Lib/TLibEncoder/TEncSbac.h	2012-08-17 16:27:37.000000000 +0900
@@ -198,7 +198,6 @@
   Void estSignificantCoefficientsBit ( estBitsSbacStruct* pcEstBitsSbac, UInt uiCTXIdx, TextType eTType );
   
   Void updateContextTables           ( SliceType eSliceType, Int iQp, Bool bExecuteFinish=true  );
-  Void updateContextTables           ( SliceType eSliceType, Int iQp  ) { this->updateContextTables( eSliceType, iQp, true); };
   
   TEncBinIf* getEncBinIf()  { return m_pcBinIf; }
 private:
diff -urN original/Lib/TLibEncoder/TEncSlice.cpp source/Lib/TLibEncoder/TEncSlice.cpp
--- original/Lib/TLibEncoder/TEncSlice.cpp	2012-08-07 03:11:44.000000000 +0900
+++ source/Lib/TLibEncoder/TEncSlice.cpp	2012-08-22 21:38:48.000000000 +0900
@@ -805,7 +805,7 @@
        uiCUAddr = rpcPic->getPicSym()->getCUOrderMap(++uiEncCUOrder) )
   {
     // initialize CU encoder
-    TComDataCU*& pcCU = rpcPic->getCU( uiCUAddr );
+    TComDataCU* pcCU = rpcPic->getCU( uiCUAddr );
     pcCU->initCU( rpcPic, uiCUAddr );
 
     if(m_pcCfg->getUseRateCtrl())
@@ -1235,7 +1235,7 @@
       }
     }
 
-    TComDataCU*& pcCU = rpcPic->getCU( uiCUAddr );    
+    TComDataCU* pcCU = rpcPic->getCU( uiCUAddr );    
 #if !SAO_LUM_CHROMA_ONOFF_FLAGS
     if ( pcSlice->getSPS()->getUseSAO() && pcSlice->getSaoEnabledFlag() )
 #else
